<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.2.313">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>2025 Coriell Bioinformatics Internship - 2&nbsp; R bootcamp</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1.6em;
  vertical-align: middle;
}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { color: #008000; } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { color: #008000; font-weight: bold; } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./data-viz.html" rel="next">
<link href="./data-best-practices.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
    <div class="container-fluid d-flex justify-content-between">
      <h1 class="quarto-secondary-nav-title"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R bootcamp</span></h1>
      <button type="button" class="quarto-btn-toggle btn" aria-label="Show secondary navigation">
        <i class="bi bi-chevron-right"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">2025 Coriell Bioinformatics Internship</a> 
    </div>
      </div>
      <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
      </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">Welcome!</a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-best-practices.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Some data best practices</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-basics.html" class="sidebar-item-text sidebar-link active"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R bootcamp</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-viz.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Effective data visualization</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./stats-by-simulation.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Statistics using simulations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./high-dimensional-data-analysis.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">High-dimensional data analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./presenting-science.html" class="sidebar-item-text sidebar-link"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Effective science communication</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">References</a>
  </div>
</li>
    </ul>
    </div>
</nav>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction-to-programming-in-r" id="toc-introduction-to-programming-in-r" class="nav-link active" data-scroll-target="#introduction-to-programming-in-r"><span class="toc-section-number">2.1</span>  Introduction to programming in R</a></li>
  <li><a href="#subsetting-in-r" id="toc-subsetting-in-r" class="nav-link" data-scroll-target="#subsetting-in-r"><span class="toc-section-number">2.2</span>  Subsetting in R</a>
  <ul class="collapse">
  <li><a href="#subsetting-vectors" id="toc-subsetting-vectors" class="nav-link" data-scroll-target="#subsetting-vectors"><span class="toc-section-number">2.2.1</span>  Subsetting vectors</a></li>
  <li><a href="#subsetting-data.frames" id="toc-subsetting-data.frames" class="nav-link" data-scroll-target="#subsetting-data.frames"><span class="toc-section-number">2.2.2</span>  Subsetting data.frames</a></li>
  <li><a href="#subsetting-matrices" id="toc-subsetting-matrices" class="nav-link" data-scroll-target="#subsetting-matrices"><span class="toc-section-number">2.2.3</span>  Subsetting matrices</a></li>
  </ul></li>
  <li><a href="#what-are-summarizedexperiments" id="toc-what-are-summarizedexperiments" class="nav-link" data-scroll-target="#what-are-summarizedexperiments"><span class="toc-section-number">2.3</span>  What are <code>SummarizedExperiments</code></a>
  <ul class="collapse">
  <li><a href="#constructing-a-summarizedexperiment" id="toc-constructing-a-summarizedexperiment" class="nav-link" data-scroll-target="#constructing-a-summarizedexperiment"><span class="toc-section-number">2.3.1</span>  Constructing a <code>SummarizedExperiment</code></a></li>
  </ul></li>
  <li><a href="#accessing-parts-of-the-summarizedexperiment-object" id="toc-accessing-parts-of-the-summarizedexperiment-object" class="nav-link" data-scroll-target="#accessing-parts-of-the-summarizedexperiment-object"><span class="toc-section-number">2.4</span>  Accessing parts of the <code>SummarizedExperiment</code> object</a>
  <ul class="collapse">
  <li><a href="#getting-the-count-matrix" id="toc-getting-the-count-matrix" class="nav-link" data-scroll-target="#getting-the-count-matrix"><span class="toc-section-number">2.4.1</span>  Getting the count matrix</a></li>
  <li><a href="#getting-the-column-metadata" id="toc-getting-the-column-metadata" class="nav-link" data-scroll-target="#getting-the-column-metadata"><span class="toc-section-number">2.4.2</span>  Getting the column metadata</a></li>
  <li><a href="#getting-the-rowranges" id="toc-getting-the-rowranges" class="nav-link" data-scroll-target="#getting-the-rowranges"><span class="toc-section-number">2.4.3</span>  Getting the rowRanges</a></li>
  <li><a href="#getting-the-rowdata" id="toc-getting-the-rowdata" class="nav-link" data-scroll-target="#getting-the-rowdata"><span class="toc-section-number">2.4.4</span>  Getting the rowData</a></li>
  </ul></li>
  <li><a href="#modifying-a-summarizedexperiment" id="toc-modifying-a-summarizedexperiment" class="nav-link" data-scroll-target="#modifying-a-summarizedexperiment"><span class="toc-section-number">2.5</span>  Modifying a <code>SummarizedExperiment</code></a>
  <ul class="collapse">
  <li><a href="#adding-assays" id="toc-adding-assays" class="nav-link" data-scroll-target="#adding-assays"><span class="toc-section-number">2.5.1</span>  Adding assays</a></li>
  <li><a href="#adding-metadata" id="toc-adding-metadata" class="nav-link" data-scroll-target="#adding-metadata"><span class="toc-section-number">2.5.2</span>  Adding metadata</a></li>
  <li><a href="#adding-rowdata" id="toc-adding-rowdata" class="nav-link" data-scroll-target="#adding-rowdata"><span class="toc-section-number">2.5.3</span>  Adding rowData</a></li>
  </ul></li>
  <li><a href="#subsetting-summarizedexperiment-objects" id="toc-subsetting-summarizedexperiment-objects" class="nav-link" data-scroll-target="#subsetting-summarizedexperiment-objects"><span class="toc-section-number">2.6</span>  Subsetting <code>SummarizedExperiment</code> objects</a>
  <ul class="collapse">
  <li><a href="#subsetting-based-on-sample-metadata" id="toc-subsetting-based-on-sample-metadata" class="nav-link" data-scroll-target="#subsetting-based-on-sample-metadata"><span class="toc-section-number">2.6.1</span>  Subsetting based on sample metadata</a></li>
  <li><a href="#subsetting-based-on-rows" id="toc-subsetting-based-on-rows" class="nav-link" data-scroll-target="#subsetting-based-on-rows"><span class="toc-section-number">2.6.2</span>  Subsetting based on rows</a></li>
  <li><a href="#subsetting-based-on-rowranges" id="toc-subsetting-based-on-rowranges" class="nav-link" data-scroll-target="#subsetting-based-on-rowranges"><span class="toc-section-number">2.6.3</span>  Subsetting based on rowRanges</a></li>
  <li><a href="#combining-subsetting-operations" id="toc-combining-subsetting-operations" class="nav-link" data-scroll-target="#combining-subsetting-operations"><span class="toc-section-number">2.6.4</span>  Combining subsetting operations</a></li>
  </ul></li>
  <li><a href="#saving-a-summarizedexperiment" id="toc-saving-a-summarizedexperiment" class="nav-link" data-scroll-target="#saving-a-summarizedexperiment"><span class="toc-section-number">2.7</span>  Saving a <code>SummarizedExperiment</code></a></li>
  <li><a href="#summarizedexperiments-in-the-real-world" id="toc-summarizedexperiments-in-the-real-world" class="nav-link" data-scroll-target="#summarizedexperiments-in-the-real-world"><span class="toc-section-number">2.8</span>  <code>SummarizedExperiments</code> in the real world</a></li>
  <li><a href="#closing-thoughts" id="toc-closing-thoughts" class="nav-link" data-scroll-target="#closing-thoughts"><span class="toc-section-number">2.9</span>  Closing thoughts</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title d-none d-lg-block"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">R bootcamp</span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<section id="introduction-to-programming-in-r" class="level2" data-number="2.1">
<h2 data-number="2.1" class="anchored" data-anchor-id="introduction-to-programming-in-r"><span class="header-section-number">2.1</span> Introduction to programming in R</h2>
<p>There are tons of great resources for learning R. <a href="https://r4ds.had.co.nz/">R for Data Science</a> is probably the most popular resource for new useRs to get up to speed with slicing and dicing data in R. The <em>R for Data Science</em> book, however, is taught from the perspective of the <a href="https://www.tidyverse.org/">Tidyverse</a>. The <em>Tidyverse</em> is an opinionated set of packages and functions that help users perform data manipulations primarily on data.frames. While these packages and functions can be great for experienced users by providing ergonomic and consistent interfaces for data.frame manipulation, it is my personal belief that new users should first learn the base language, especially if their goal is to perform bioinformatics analysis.</p>
<p>Bioinformatics tools rely heavily on subsetting and matrix manipulations. In my experience, users who start learning R using only function from the <em>Tidyverse</em> have a difficult time understanding matrix manipulations and subsetting operations common in bioinformatics workflows. This becomes especially important when using <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiments</a> - the backbone of many bioinformatics data structures in R.</p>
<p>For this reason, we’re going to focus on learning R from the ground up using functions that exist primarily in the base language. A great resource for learning base R quickly is Norm Matloff’s <em>fasteR</em> which can be found <a href="https://github.com/matloff/fasteR">here</a>.</p>
</section>
<section id="subsetting-in-r" class="level2" data-number="2.2">
<h2 data-number="2.2" class="anchored" data-anchor-id="subsetting-in-r"><span class="header-section-number">2.2</span> Subsetting in R</h2>
<p>You may have only ever encountered R from the perspective of the <a href="https://www.tidyverse.org/">tidyverse</a>. <code>tidyverse</code> functions provide useful abstractions for munging <a href="https://vita.had.co.nz/papers/tidy-data.html">tidy data</a> however, most genomics data is often best represented and operated on as matrices. Keeping your data in matrix format can provide many benefits as far as speed and code clarity, which in turn helps to ensure correctness. You can think of matrices as just fancy 2D versions of vectors. So what are vectors?</p>
<p>Vectors are the main building blocks of most R analyses. Whenever you use the <code>c()</code> function, like: <code>x &lt;- c('a', 'b', 'c')</code> you’re creating a vector. You can do all kinds of cool things with vectors which will prove useful when working with <code>SummarizedExperiments</code>.</p>
<p>NOTE: the following is heavily inspired by Norm Matloff’s excellent <a href="https://github.com/matloff/fasteR">fasteR</a> tutorial. Take a look there to get a brief and concise overview base R. You should also check out the first few chapters of Hadley Wickham’s amazing book <a href="https://adv-r.hadley.nz/">Advanced R</a>. The <a href="http://adv-r.had.co.nz/">first edition</a> contains some more information on base R.</p>
<section id="subsetting-vectors" class="level3" data-number="2.2.1">
<h3 data-number="2.2.1" class="anchored" data-anchor-id="subsetting-vectors"><span class="header-section-number">2.2.1</span> Subsetting vectors</h3>
<p>Below, we’ll use the built-in R constant called <code>LETTERS</code>. The <code>LETTERS</code> vector is simply a ‘list’ of all uppercase letters in the Roman alphabet.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>LETTERS</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"
[20] "T" "U" "V" "W" "X" "Y" "Z"</code></pre>
</div>
</div>
<p>We can subset the vector by position. For example, to get the 3rd letter we use the <code>[</code> operator and the position we want to extract.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C"</code></pre>
</div>
</div>
<p>We can also use a range of positions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">3</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C" "D" "E" "F" "G"</code></pre>
</div>
</div>
<p>We don’t have to select sequential elements either. We can extract elements by using another vector of positions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="fu">c</span>(<span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">14</span>, <span class="dv">14</span>, <span class="dv">1</span>, <span class="dv">18</span>, <span class="dv">15</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "G" "E" "N" "N" "A" "R" "O"</code></pre>
</div>
</div>
<p>Vectors become really powerful when we start combining them with logical operations.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>my_favorite_letters <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># See that this produces a logical vector of (TRUE/FALSE) values</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># TRUE when LETTERS is one of my_favorite_letters and FALSE otherwise</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>LETTERS <span class="sc">%in%</span> my_favorite_letters</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[25] FALSE FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use that same expression to filter the vector</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>LETTERS[LETTERS <span class="sc">%in%</span> my_favorite_letters]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "B" "C"</code></pre>
</div>
</div>
<p>This same kind of subsetting works on vectors that contain numeric data as well. For example, we can filter the measurements of annual flow of water through the Nile river like so:</p>
<p><em>Nile is another built-in dataset</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Any values strictly greater than 1200</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>Nile[Nile <span class="sc">&gt;</span> <span class="dv">1200</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1210 1230 1370 1210 1250 1260 1220</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Any even number</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>Nile[Nile <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">==</span> <span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1120 1160 1210 1160 1160 1230 1370 1140 1110  994 1020  960 1180  958 1140
[16] 1100 1210 1150 1250 1260 1220 1030 1100  774  840  874  694  940  916  692
[31] 1020 1050  726  456  824  702 1120 1100  832  764  768  864  862  698  744
[46]  796 1040  944  984  822 1010  676  846  812  742 1040  860  874  848  890
[61]  744  838 1050  918  986 1020  906 1170  912  746  718  714  740</code></pre>
</div>
</div>
</section>
<section id="subsetting-data.frames" class="level3" data-number="2.2.2">
<h3 data-number="2.2.2" class="anchored" data-anchor-id="subsetting-data.frames"><span class="header-section-number">2.2.2</span> Subsetting data.frames</h3>
<p>But these are just one dimensional vectors. In R we usually deal with data.frames (tibbles for you tidyverse) and matrices. Lucky for us, the subsetting operations we learned for vectors work the same way for data.frames and matrices.</p>
<p>Let’s take a look at the built-in <code>ToothGrowth</code> dataset. The data consists of the length of odontoblasts in 60 guinea pigs receiving one of three levels of vitamin C by one of two delivery methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
1  4.2   VC  0.5
2 11.5   VC  0.5
3  7.3   VC  0.5
4  5.8   VC  0.5
5  6.4   VC  0.5
6 10.0   VC  0.5</code></pre>
</div>
</div>
<p>The dollar sign <code>$</code> is used to extract an individual column from the data.frame, which is just a vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth<span class="sc">$</span>len)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  4.2 11.5  7.3  5.8  6.4 10.0</code></pre>
</div>
</div>
<p>We can also use the <code>[[</code> to get the same thing. Double-brackets come in handy when your columns are not valid R names since <code>$</code> only works when columns are valid names</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth[[<span class="st">"len"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  4.2 11.5  7.3  5.8  6.4 10.0</code></pre>
</div>
</div>
<p>When subsetting a data.frame in base R the general scheme is:</p>
<pre><code>df[the rows you want, the columns you want]</code></pre>
<p>So in order to get the 5th row of the first column we could do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[<span class="dv">5</span>, <span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6.4</code></pre>
</div>
</div>
<p>Again, we can combine this kind of thinking to extract rows and columns matching logical conditions. For example, if we want to get all of the animals administered orange juice (‘OJ’)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    len supp dose
31 15.2   OJ  0.5
32 21.5   OJ  0.5
33 17.6   OJ  0.5
34  9.7   OJ  0.5
35 14.5   OJ  0.5
36 10.0   OJ  0.5
37  8.2   OJ  0.5
38  9.4   OJ  0.5
39 16.5   OJ  0.5
40  9.7   OJ  0.5
41 19.7   OJ  1.0
42 23.3   OJ  1.0
43 23.6   OJ  1.0
44 26.4   OJ  1.0
45 20.0   OJ  1.0
46 25.2   OJ  1.0
47 25.8   OJ  1.0
48 21.2   OJ  1.0
49 14.5   OJ  1.0
50 27.3   OJ  1.0
51 25.5   OJ  2.0
52 26.4   OJ  2.0
53 22.4   OJ  2.0
54 24.5   OJ  2.0
55 24.8   OJ  2.0
56 30.9   OJ  2.0
57 26.4   OJ  2.0
58 27.3   OJ  2.0
59 29.4   OJ  2.0
60 23.0   OJ  2.0</code></pre>
</div>
</div>
<p>We can also combine logical statements. For example, to get all of the rows for animals administered orange juice and with odontoblast length (‘len’) less than 10.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> ToothGrowth<span class="sc">$</span>len <span class="sc">&lt;</span> <span class="dv">10</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
34 9.7   OJ  0.5
37 8.2   OJ  0.5
38 9.4   OJ  0.5
40 9.7   OJ  0.5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can also use the bracket notation to select rows and columns at the same time</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> ToothGrowth<span class="sc">$</span>len <span class="sc">&lt;</span> <span class="dv">10</span>, <span class="fu">c</span>(<span class="st">"len"</span>, <span class="st">"supp"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp
34 9.7   OJ
37 8.2   OJ
38 9.4   OJ
40 9.7   OJ</code></pre>
</div>
</div>
<p>It gets annoying typing <code>ToothGrowth</code> every time we want to subset the data.frame. Base R has a very useful function called <code>subset()</code> that can help us type less. <code>subset()</code> essentially ‘looks inside’ the data.frame that you give it for the given columns and evaluates the expression without having to explicitly tell R where to find the columns. Think of it like <code>dplyr::filter()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(ToothGrowth, supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> len <span class="sc">&lt;</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
34 9.7   OJ  0.5
37 8.2   OJ  0.5
38 9.4   OJ  0.5
40 9.7   OJ  0.5</code></pre>
</div>
</div>
</section>
<section id="subsetting-matrices" class="level3" data-number="2.2.3">
<h3 data-number="2.2.3" class="anchored" data-anchor-id="subsetting-matrices"><span class="header-section-number">2.2.3</span> Subsetting matrices</h3>
<p>Matrices behave much like data.frames but unlike data.frames matrices can only contain one type of data. This might sound like a limitation at first but you’ll soon come to realize that matrices are very powerful (and fast) to work with in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create some random data that looks like methylation values</span></span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>(m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">runif</span>(<span class="dv">6</span> <span class="sc">*</span> <span class="dv">10</span>),</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">6</span>,</span>
<span id="cb34-7"><a href="#cb34-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(</span>
<span id="cb34-8"><a href="#cb34-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"CpG."</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb34-9"><a href="#cb34-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"Sample"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb34-10"><a href="#cb34-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb34-11"><a href="#cb34-11" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Sample1    Sample2   Sample3    Sample4   Sample5    Sample6
CpG.1  0.2875775 0.95683335 0.8895393 0.96302423 0.1428000 0.04583117
CpG.2  0.7883051 0.45333416 0.6928034 0.90229905 0.4145463 0.44220007
CpG.3  0.4089769 0.67757064 0.6405068 0.69070528 0.4137243 0.79892485
CpG.4  0.8830174 0.57263340 0.9942698 0.79546742 0.3688455 0.12189926
CpG.5  0.9404673 0.10292468 0.6557058 0.02461368 0.1524447 0.56094798
CpG.6  0.0455565 0.89982497 0.7085305 0.47779597 0.1388061 0.20653139
CpG.7  0.5281055 0.24608773 0.5440660 0.75845954 0.2330341 0.12753165
CpG.8  0.8924190 0.04205953 0.5941420 0.21640794 0.4659625 0.75330786
CpG.9  0.5514350 0.32792072 0.2891597 0.31818101 0.2659726 0.89504536
CpG.10 0.4566147 0.95450365 0.1471136 0.23162579 0.8578277 0.37446278</code></pre>
</div>
</div>
<p>If we want to extract the value for CpG.3 for Sample3</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">3</span>, <span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.6405068</code></pre>
</div>
</div>
<p>Or all values of CpG.3 for every sample</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">3</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample1   Sample2   Sample3   Sample4   Sample5   Sample6 
0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or refer to the row by it's name</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>m[<span class="st">"CpG.3"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample1   Sample2   Sample3   Sample4   Sample5   Sample6 
0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248 </code></pre>
</div>
</div>
<p>Or all CpGs for Sample3</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>m[, <span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    CpG.1     CpG.2     CpG.3     CpG.4     CpG.5     CpG.6     CpG.7     CpG.8 
0.8895393 0.6928034 0.6405068 0.9942698 0.6557058 0.7085305 0.5440660 0.5941420 
    CpG.9    CpG.10 
0.2891597 0.1471136 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or refer to the column by it's name</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>m[, <span class="st">"Sample3"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    CpG.1     CpG.2     CpG.3     CpG.4     CpG.5     CpG.6     CpG.7     CpG.8 
0.8895393 0.6928034 0.6405068 0.9942698 0.6557058 0.7085305 0.5440660 0.5941420 
    CpG.9    CpG.10 
0.2891597 0.1471136 </code></pre>
</div>
</div>
<p>We can also apply a mask to the entire matrix at once. For example, the following will mark any value that is greater than 0.5 with <code>TRUE</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>m <span class="sc">&gt;</span> <span class="fl">0.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
CpG.1    FALSE    TRUE    TRUE    TRUE   FALSE   FALSE
CpG.2     TRUE   FALSE    TRUE    TRUE   FALSE   FALSE
CpG.3    FALSE    TRUE    TRUE    TRUE   FALSE    TRUE
CpG.4     TRUE    TRUE    TRUE    TRUE   FALSE   FALSE
CpG.5     TRUE   FALSE    TRUE   FALSE   FALSE    TRUE
CpG.6    FALSE    TRUE    TRUE   FALSE   FALSE   FALSE
CpG.7     TRUE   FALSE    TRUE    TRUE   FALSE   FALSE
CpG.8     TRUE   FALSE    TRUE   FALSE   FALSE    TRUE
CpG.9     TRUE   FALSE   FALSE   FALSE   FALSE    TRUE
CpG.10   FALSE    TRUE   FALSE   FALSE    TRUE   FALSE</code></pre>
</div>
</div>
<p>We can use this kind of masking to filter rows of the matrix using some very helpful base R functions that operate on matrices. For example, to get only those CpGs where 3 or more samples have a value &gt; 0.5 we can use the <code>rowSums()</code> like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>m[<span class="fu">rowSums</span>(m <span class="sc">&gt;</span> <span class="fl">0.5</span>) <span class="sc">&gt;</span> <span class="dv">3</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1   Sample2   Sample3   Sample4   Sample5   Sample6
CpG.3 0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248
CpG.4 0.8830174 0.5726334 0.9942698 0.7954674 0.3688455 0.1218993</code></pre>
</div>
</div>
<p>This pattern is very common when dealing with sequencing data. Base R functions like <code>rowSums()</code> and <code>colMeans()</code> are specialized to operate over matrices and are the most efficient way to summarize matrix data. The R package <a href="https://github.com/HenrikBengtsson/matrixStats">matrixStats</a> also contains highly optimized functions for operating on matrices.</p>
<p>Compare the above to the <code>tidy</code> solution given the same matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>tidyr<span class="sc">::</span><span class="fu">as_tibble</span>(m, <span class="at">rownames =</span> <span class="st">"CpG"</span>) <span class="sc">|&gt;</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">!</span>CpG, <span class="at">names_to =</span> <span class="st">"SampleName"</span>, <span class="at">values_to =</span> <span class="st">"beta"</span>) <span class="sc">|&gt;</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(CpG) <span class="sc">|&gt;</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n =</span> <span class="fu">sum</span>(beta <span class="sc">&gt;</span> <span class="fl">0.5</span>)) <span class="sc">|&gt;</span></span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(n <span class="sc">&gt;</span> <span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> CpG, <span class="at">names_from =</span> <span class="st">"SampleName"</span>, <span class="at">values_from =</span> <span class="st">"beta"</span>) <span class="sc">|&gt;</span></span>
<span id="cb50-7"><a href="#cb50-7" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"CpG"</span>) <span class="sc">|&gt;</span></span>
<span id="cb50-8"><a href="#cb50-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1   Sample2   Sample3   Sample4   Sample5   Sample6
CpG.3 0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248
CpG.4 0.8830174 0.5726334 0.9942698 0.7954674 0.3688455 0.1218993</code></pre>
</div>
</div>
<p><em>There’s probably some kind of tidy solution using <code>across()</code> that I’m missing but this is how most of the tidy code in the wild that I have seen looks</em></p>
<p>Hopefully these examples are enough to get you started understanding how subsetting works in R and appreciate how useful it is. Now that you have some familiarity with the using R functions for subsetting objects, you’re ready to start working with <code>SummarizedExperiments</code>.</p>
</section>
</section>
<section id="what-are-summarizedexperiments" class="level2" data-number="2.3">
<h2 data-number="2.3" class="anchored" data-anchor-id="what-are-summarizedexperiments"><span class="header-section-number">2.3</span> What are <code>SummarizedExperiments</code></h2>
<p><code>SummarizedExperiments</code> are R objects meant for organizing and manipulating rectangular matrices that are typically produced by arrays or high-throughput sequencing. If you are doing any kind of analysis that requires associating feature-level data (RNA-seq gene counts, methylation array loci, ATAC-seq regions, etc.) with the genomic coordinates of those features and the sample-level metadata with which those features were measured, then you should be using a <code>SummarizedExperiment</code> to organize, manipulate, and store your results.</p>
<p>Please take a moment to read through the first 2 sections (at least) of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment vignette</a> in order to familiarize yourself with what <code>SummarizedExperiments</code> are and their structure. I will demonstrate <em>how</em> you can use <code>SummarizedExperiments</code> below.</p>
<p>From the <code>SummarizedExperiment</code> <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">vignette</a>:</p>
<p><img src="images/summarized-experiments.png" title="Overview of SummarizedExperiment structure" class="img-fluid"></p>
<p>The <code>SummarizedExperiment</code> object coordinates four main parts:</p>
<ul>
<li><code>assay()</code>, <code>assays()</code>: A matrix-like or list of matrix-like objects of identical dimension
<ul>
<li>matrix-like: implements <code>dim()</code>, <code>dimnames()</code>, and 2-dimensional <code>[</code>, <code>[&lt;-</code> methods.</li>
<li>rows: genes, genomic coordinates, etc.</li>
<li>columns: samples, cells, etc.</li>
</ul></li>
<li><code>colData()</code>: Annotations on each column, as a DataFrame.
<ul>
<li>E.g., description of each sample</li>
</ul></li>
<li><code>rowData()</code> and / or <code>rowRanges()</code>: Annotations on each row.
<ul>
<li><code>rowRanges()</code>: coordinates of gene / exons in transcripts / etc.</li>
<li><code>rowData()</code>: P-values and log-fold change of each gene after differential expression analysis.</li>
</ul></li>
<li><code>metadata()</code>: List of unstructured metadata describing the overall content of the object.</li>
</ul>
<p>In order to better understand how they work, let’s construct a <code>SummarizedExperiment</code> from scratch.</p>
<section id="constructing-a-summarizedexperiment" class="level3" data-number="2.3.1">
<h3 data-number="2.3.1" class="anchored" data-anchor-id="constructing-a-summarizedexperiment"><span class="header-section-number">2.3.1</span> Constructing a <code>SummarizedExperiment</code></h3>
<p>Hopefully you’ll already be working with data that is in a <code>SummarizedExperiment</code> or some other class that derives from one. But just in case you don’t have data structured as a <code>SummarizedExperiment</code> it’s useful and instructive to understand how to create one from scratch.</p>
<p>To be most useful, a <code>SummarizedExperiment</code> should have at least:</p>
<ul>
<li>A matrix of data with features in rows and samples in columns</li>
<li>A metadata data.frame with samples as rownames and columns describing their properties</li>
</ul>
<p>Another really useful object to add to <code>SummarizedExperiments</code> is a GRanges object describing the genomic locations of each feature in the matrix. Adding this to the <code>SummarizedExperiment</code> creates what is called a <code>RangedSummarizedExperiment</code> that acts just like a regular <code>SummarizedExperiment</code> with some extra features.</p>
<p>To construct our basic <code>SummarizedExperiment</code>:</p>
<ul>
<li>We’ll create a ‘counts’ matrix with gene IDs as rows and Samples in columns</li>
<li>We’ll add some metadata describing the Samples</li>
<li>We’ll add on <code>GRanges()</code> describing the genomic location of the genes</li>
</ul>
<section id="construct-the-counts-matrix" class="level4" data-number="2.3.1.1">
<h4 data-number="2.3.1.1" class="anchored" data-anchor-id="construct-the-counts-matrix"><span class="header-section-number">2.3.1.1</span> Construct the counts matrix</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(SummarizedExperiment))</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">rnbinom</span>(<span class="at">n =</span> <span class="dv">200</span> <span class="sc">*</span> <span class="dv">6</span>, <span class="at">mu =</span> <span class="dv">100</span>, <span class="at">size =</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fl">0.5</span>),</span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">200</span>,</span>
<span id="cb52-7"><a href="#cb52-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="fu">paste0</span>(<span class="st">"Gene"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>), <span class="fu">paste0</span>(<span class="st">"Sample"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>))</span>
<span id="cb52-8"><a href="#cb52-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb52-9"><a href="#cb52-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-10"><a href="#cb52-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb52-11"><a href="#cb52-11" aria-hidden="true" tabindex="-1"></a>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2 Sample3 Sample4 Sample5
Gene1     100     210     186      14      95
Gene2      74      17      62      48      27
Gene3     129      72     105     203      73
Gene4      73      80      84      81      59
Gene5      17     242      32      21      58</code></pre>
</div>
</div>
</section>
<section id="construct-the-sample-metadata" class="level4" data-number="2.3.1.2">
<h4 data-number="2.3.1.2" class="anchored" data-anchor-id="construct-the-sample-metadata"><span class="header-section-number">2.3.1.2</span> Construct the sample metadata</h4>
<p>It is important that the sample metadata be either a <code>data.frame</code> or a <code>DataFrame</code> object because <code>SummarizedExperiment</code> requires the <code>colData()</code> to have rownames that match the colnames of the count matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">SampleName =</span> <span class="fu">colnames</span>(counts),</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Treatment =</span> <span class="fu">gl</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Control"</span>, <span class="st">"Treatment"</span>)),</span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Age =</span> <span class="fu">sample.int</span>(<span class="dv">100</span>, <span class="dv">6</span>),</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="fu">colnames</span>(counts)</span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-8"><a href="#cb54-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb54-9"><a href="#cb54-9" aria-hidden="true" tabindex="-1"></a>coldata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        SampleName Treatment Age
Sample1    Sample1   Control  38
Sample2    Sample2   Control  71
Sample3    Sample3   Control  30
Sample4    Sample4 Treatment  64
Sample5    Sample5 Treatment   8
Sample6    Sample6 Treatment  47</code></pre>
</div>
</div>
<p>Notice that all of the rownames of the metadata are in the same order as the colnames of the counts matrix. <strong>This is necessary</strong>.</p>
</section>
<section id="construct-gene-range-annotations" class="level4" data-number="2.3.1.3">
<h4 data-number="2.3.1.3" class="anchored" data-anchor-id="construct-gene-range-annotations"><span class="header-section-number">2.3.1.3</span> Construct gene range annotations</h4>
<p>You will usually have gene annotations or <code>GRanges</code> objects loaded from a GTF file or you may even create <code>GRanges</code> yourself by specifying the chromosome, start, end, and strand, information manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>rowranges <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="fu">c</span>(<span class="dv">50</span>, <span class="dv">150</span>)),</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">IRanges</span>(<span class="fu">floor</span>(<span class="fu">runif</span>(<span class="dv">200</span>, <span class="fl">1e5</span>, <span class="fl">1e6</span>)), <span class="at">width =</span> <span class="dv">100</span>),</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">strand =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">200</span>, <span class="cn">TRUE</span>),</span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">feature_id =</span> <span class="fu">sprintf</span>(<span class="st">"ID%03d"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>),</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_type =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"protein_coding"</span>, <span class="st">"lncRNA"</span>, <span class="st">"repeat_element"</span>), <span class="dv">200</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb56-7"><a href="#cb56-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb56-8"><a href="#cb56-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(rowranges) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb56-9"><a href="#cb56-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb56-10"><a href="#cb56-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb56-11"><a href="#cb56-11" aria-hidden="true" tabindex="-1"></a>rowranges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 2 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt;
    Gene1     chr1 954187-954286      - |       ID001 protein_coding
    Gene2     chr1 552363-552462      + |       ID002 protein_coding
    Gene3     chr1 303793-303892      - |       ID003         lncRNA
    Gene4     chr1 875980-876079      + |       ID004 protein_coding
    Gene5     chr1 427141-427240      - |       ID005 repeat_element
      ...      ...           ...    ... .         ...            ...
  Gene196     chr2 268871-268970      + |       ID196 repeat_element
  Gene197     chr2 599914-600013      + |       ID197 repeat_element
  Gene198     chr2 477464-477563      + |       ID198 repeat_element
  Gene199     chr2 461424-461523      - |       ID199 repeat_element
  Gene200     chr2 619939-620038      + |       ID200 protein_coding
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="construct-the-summarizedexperiment-object" class="level4" data-number="2.3.1.4">
<h4 data-number="2.3.1.4" class="anchored" data-anchor-id="construct-the-summarizedexperiment-object"><span class="header-section-number">2.3.1.4</span> Construct the <code>SummarizedExperiment</code> object</h4>
<p>With these pieces of information we’re ready to create a <code>SummarizedExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts),</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> coldata,</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowRanges =</span> rowranges</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Printing the object gives a summary of what's inside</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 6 
metadata(0):
assays(1): counts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(2): feature_id gene_type
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(3): SampleName Treatment Age</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="accessing-parts-of-the-summarizedexperiment-object" class="level2" data-number="2.4">
<h2 data-number="2.4" class="anchored" data-anchor-id="accessing-parts-of-the-summarizedexperiment-object"><span class="header-section-number">2.4</span> Accessing parts of the <code>SummarizedExperiment</code> object</h2>
<p>Every part of the <code>SummarizedExperiment</code> object can be extracted with its accessor function. To extract a particular assay you can use the <code>assay()</code> function. To extract the column metadata you can use the <code>colData()</code> function. To extract the GRanges for the rows of the matrix you can use the <code>rowRanges()</code> function. The <code>rowData()</code> function also allows you to access row-level annotation information from data added to the <code>rowData</code> slot or by the <code>mcols()</code> of the <code>rowRanges</code>. This will be made more clear below.</p>
<section id="getting-the-count-matrix" class="level3" data-number="2.4.1">
<h3 data-number="2.4.1" class="anchored" data-anchor-id="getting-the-count-matrix"><span class="header-section-number">2.4.1</span> Getting the count matrix</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"counts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
Gene1     100     210     186      14      95      43
Gene2      74      17      62      48      27      71
Gene3     129      72     105     203      73      63
Gene4      73      80      84      81      59     153
Gene5      17     242      32      21      58      40
Gene6      22      91      38      89     164      43</code></pre>
</div>
</div>
<p>To see what assays are available you can use the <code>assays()</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 1
names(1): counts</code></pre>
</div>
</div>
</section>
<section id="getting-the-column-metadata" class="level3" data-number="2.4.2">
<h3 data-number="2.4.2" class="anchored" data-anchor-id="getting-the-column-metadata"><span class="header-section-number">2.4.2</span> Getting the column metadata</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
         SampleName Treatment       Age
        &lt;character&gt;  &lt;factor&gt; &lt;integer&gt;
Sample1     Sample1 Control          38
Sample2     Sample2 Control          71
Sample3     Sample3 Control          30
Sample4     Sample4 Treatment        64
Sample5     Sample5 Treatment         8
Sample6     Sample6 Treatment        47</code></pre>
</div>
</div>
</section>
<section id="getting-the-rowranges" class="level3" data-number="2.4.3">
<h3 data-number="2.4.3" class="anchored" data-anchor-id="getting-the-rowranges"><span class="header-section-number">2.4.3</span> Getting the rowRanges</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 2 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt;
    Gene1     chr1 954187-954286      - |       ID001 protein_coding
    Gene2     chr1 552363-552462      + |       ID002 protein_coding
    Gene3     chr1 303793-303892      - |       ID003         lncRNA
    Gene4     chr1 875980-876079      + |       ID004 protein_coding
    Gene5     chr1 427141-427240      - |       ID005 repeat_element
      ...      ...           ...    ... .         ...            ...
  Gene196     chr2 268871-268970      + |       ID196 repeat_element
  Gene197     chr2 599914-600013      + |       ID197 repeat_element
  Gene198     chr2 477464-477563      + |       ID198 repeat_element
  Gene199     chr2 461424-461523      - |       ID199 repeat_element
  Gene200     chr2 619939-620038      + |       ID200 protein_coding
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="getting-the-rowdata" class="level3" data-number="2.4.4">
<h3 data-number="2.4.4" class="anchored" data-anchor-id="getting-the-rowdata"><span class="header-section-number">2.4.4</span> Getting the rowData</h3>
<p>Note that rowData in this case is the same as <code>mcols()</code> of the <code>rowRanges</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 200 rows and 2 columns
         feature_id      gene_type
        &lt;character&gt;    &lt;character&gt;
Gene1         ID001 protein_coding
Gene2         ID002 protein_coding
Gene3         ID003         lncRNA
Gene4         ID004 protein_coding
Gene5         ID005 repeat_element
...             ...            ...
Gene196       ID196 repeat_element
Gene197       ID197 repeat_element
Gene198       ID198 repeat_element
Gene199       ID199 repeat_element
Gene200       ID200 protein_coding</code></pre>
</div>
</div>
</section>
</section>
<section id="modifying-a-summarizedexperiment" class="level2" data-number="2.5">
<h2 data-number="2.5" class="anchored" data-anchor-id="modifying-a-summarizedexperiment"><span class="header-section-number">2.5</span> Modifying a <code>SummarizedExperiment</code></h2>
<p>Once you create a <code>SummarizedExperiment</code> you are not stuck with the information in that object. <code>SummarizedExperiments</code> allow you to add and modify the data within the object.</p>
<section id="adding-assays" class="level3" data-number="2.5.1">
<h3 data-number="2.5.1" class="anchored" data-anchor-id="adding-assays"><span class="header-section-number">2.5.1</span> Adding assays</h3>
<p>For example, we may wish to calculate counts per million values for our counts matrix and add a new assay back into our <code>SummarizedExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate counts per million</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(se, <span class="st">"counts"</span>)</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>cpm <span class="ot">&lt;-</span> counts <span class="sc">/</span> <span class="fu">colSums</span>(counts) <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the CPM data as a new assay to our existing se object</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> cpm</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a><span class="co"># And if we wish to log-scale these values</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"logcounts"</span>) <span class="ot">&lt;-</span> <span class="fu">log2</span>(cpm)</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Now there are three assays available</span></span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 3
names(3): counts cpm logcounts</code></pre>
</div>
</div>
<p>Note: Instead of creating intermediate variables we could also directly use the assays like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(se, <span class="st">"counts"</span>) <span class="sc">/</span> <span class="fu">colSums</span>(<span class="fu">assay</span>(se, <span class="st">"counts"</span>)) <span class="sc">*</span> <span class="fl">1e6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-metadata" class="level3" data-number="2.5.2">
<h3 data-number="2.5.2" class="anchored" data-anchor-id="adding-metadata"><span class="header-section-number">2.5.2</span> Adding metadata</h3>
<p><code>SummarizedExperiment</code> objects use the <code>$</code> to get and set columns of the metadata contained in the <code>colData</code> slot. For example, to get all of the Ages we can use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>se<span class="sc">$</span>Age</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 38 71 30 64  8 47</code></pre>
</div>
</div>
<p>If we want to add a new column we simply create the new column in the same way</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>se<span class="sc">$</span>Batch <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>), <span class="dv">2</span>))</span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb75-3"><a href="#cb75-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Now you can se that a new 'Batch` column has been added to the colData</span></span>
<span id="cb75-4"><a href="#cb75-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
         SampleName Treatment       Age    Batch
        &lt;character&gt;  &lt;factor&gt; &lt;integer&gt; &lt;factor&gt;
Sample1     Sample1 Control          38        A
Sample2     Sample2 Control          71        B
Sample3     Sample3 Control          30        C
Sample4     Sample4 Treatment        64        A
Sample5     Sample5 Treatment         8        B
Sample6     Sample6 Treatment        47        C</code></pre>
</div>
</div>
</section>
<section id="adding-rowdata" class="level3" data-number="2.5.3">
<h3 data-number="2.5.3" class="anchored" data-anchor-id="adding-rowdata"><span class="header-section-number">2.5.3</span> Adding rowData</h3>
<p>We can also modify the data which describes each feature in the matrix by adding columns to the rowData. For example, let’s create a new column called <code>Keep</code> if the gene is a protein_coding gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)<span class="sc">$</span>Keep <span class="ot">&lt;-</span> <span class="fu">rowData</span>(se)<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"protein_coding"</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 200 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene1         ID001 protein_coding      TRUE
Gene2         ID002 protein_coding      TRUE
Gene3         ID003         lncRNA     FALSE
Gene4         ID004 protein_coding      TRUE
Gene5         ID005 repeat_element     FALSE
...             ...            ...       ...
Gene196       ID196 repeat_element     FALSE
Gene197       ID197 repeat_element     FALSE
Gene198       ID198 repeat_element     FALSE
Gene199       ID199 repeat_element     FALSE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
</section>
</section>
<section id="subsetting-summarizedexperiment-objects" class="level2" data-number="2.6">
<h2 data-number="2.6" class="anchored" data-anchor-id="subsetting-summarizedexperiment-objects"><span class="header-section-number">2.6</span> Subsetting <code>SummarizedExperiment</code> objects</h2>
<p><code>SummarizedExperiments</code> follow the basic idea of</p>
<pre><code>se[the rows you want, the columns you want]</code></pre>
<p>With a <code>SummarizedExperiment</code> “the rows you want” corresponds to the features in the rows of the matrix/rowData and “the columns you want” corresponds to the metadata in <code>colData</code></p>
<section id="subsetting-based-on-sample-metadata" class="level3" data-number="2.6.1">
<h3 data-number="2.6.1" class="anchored" data-anchor-id="subsetting-based-on-sample-metadata"><span class="header-section-number">2.6.1</span> Subsetting based on sample metadata</h3>
<p>For example, if we want to select all of the data belonging only to samples in the Treatment group we can use the following:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>(trt <span class="ot">&lt;-</span> se[, se<span class="sc">$</span>Treatment <span class="sc">==</span> <span class="st">"Treatment"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 3 
metadata(0):
assays(3): counts cpm logcounts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(3): Sample4 Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>Notice how the <code>dim</code> of the object changed from 6 to 3. This is because we have selected only the Samples from the original <code>SummarizedExperiment</code> object from the treatment group. The cool thing about <code>SummarizedExperiments</code> is that all of the assays have also been subsetted to reflect this selection!</p>
<p>Take a look at the “logcounts” assay. It only contains Samples 4, 5, and 6.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(trt, <span class="st">"logcounts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample4  Sample5  Sample6
Gene1  9.681439 12.23369 11.09368
Gene2 11.306469 10.37294 11.82069
Gene3 13.329167 11.85724 11.85136
Gene4 11.957898 11.55359 12.97889
Gene5 10.059737 11.73207 10.98576
Gene6 12.146676 13.07906 11.04431</code></pre>
</div>
</div>
<p>Of course you can combine multiple conditions as well</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a>se[, se<span class="sc">$</span>Batch <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"B"</span>, <span class="st">"C"</span>) <span class="sc">&amp;</span> se<span class="sc">$</span>Age <span class="sc">&gt;</span> <span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 3 
metadata(0):
assays(3): counts cpm logcounts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(3): Sample2 Sample3 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
</section>
<section id="subsetting-based-on-rows" class="level3" data-number="2.6.2">
<h3 data-number="2.6.2" class="anchored" data-anchor-id="subsetting-based-on-rows"><span class="header-section-number">2.6.2</span> Subsetting based on rows</h3>
<p>We can also select certain features that we want to keep using row subsetting. For example to select only the first 50 rows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a>se[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 50 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(50): Gene1 Gene2 ... Gene49 Gene50
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>Notice how the <code>dim</code> changed from 200 to 50 reflecting the fact that we have only selected the first 50 rows.</p>
<p>This subsetting is very useful when combined with logical operators. Above we created a vector in <code>rowData</code> called “Keep” that contained TRUE if the corresponding row of the se object was a coding gene and FALSE otherwise. Let’s use this vector to subset our se object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a>(coding <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>Keep, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 68 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(68): Gene1 Gene2 ... Gene193 Gene200
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>And if we look at the resulting <code>rowData</code> we can see that it only contains the protein_coding features</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(coding)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 68 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene1         ID001 protein_coding      TRUE
Gene2         ID002 protein_coding      TRUE
Gene4         ID004 protein_coding      TRUE
Gene8         ID008 protein_coding      TRUE
Gene13        ID013 protein_coding      TRUE
...             ...            ...       ...
Gene183       ID183 protein_coding      TRUE
Gene185       ID185 protein_coding      TRUE
Gene187       ID187 protein_coding      TRUE
Gene193       ID193 protein_coding      TRUE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
<p>Each assay also reflects this operation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(coding, <span class="st">"cpm"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1    Sample2   Sample3    Sample4  Sample5   Sample6
Gene1  5865.103 10646.3878  9453.141   821.1144 4816.223  2185.404
Gene2  3904.601   834.9295  3158.753  2532.7142 1326.065  3617.281
Gene4  3585.286  4075.8101  4432.250  3978.1936 3005.910  8073.027
Gene8  1582.946 12180.1483  1069.900  3165.8928 2603.016   815.162
Gene13 5571.848  4512.0406  6556.211 12199.4135 9936.629 15704.411
Gene15 5019.011  3405.1637 15366.569  6235.7414 4421.630  3988.270</code></pre>
</div>
</div>
</section>
<section id="subsetting-based-on-rowranges" class="level3" data-number="2.6.3">
<h3 data-number="2.6.3" class="anchored" data-anchor-id="subsetting-based-on-rowranges"><span class="header-section-number">2.6.3</span> Subsetting based on rowRanges</h3>
<p>A closely related row-wise subsetting operation can be used if you have a <code>RangedSummarizedExperiment</code> (a <code>SummarizedExperiment</code> with a rowRanges slot) that allows you to perform operations on a <code>SummarizedExperiment</code> object like you would a GRanges object.</p>
<p>For example, let’s say we only wanted to extract the features on Chromosome 2 only. Then we can use the <code>GenomicRanges</code> function <code>subsetByOverlaps</code> directly on our <code>SummarizedExperiment</code> object like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Region of interest</span></span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">"chr2"</span>, <span class="at">ranges =</span> <span class="dv">1</span><span class="sc">:</span><span class="fl">1e7</span>)</span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb94-4"><a href="#cb94-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the SE object for only features on chr2</span></span>
<span id="cb94-5"><a href="#cb94-5" aria-hidden="true" tabindex="-1"></a>(chr2 <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(se, roi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 150 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(150): Gene51 Gene52 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>You can see again that the <code>dim</code> changed reflecting our selection. Again, all of the associated assays and rowData have also been subsetted reflecting this change as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb96"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb96-1"><a href="#cb96-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(chr2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 150 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene51        ID051 protein_coding      TRUE
Gene52        ID052         lncRNA     FALSE
Gene53        ID053 repeat_element     FALSE
Gene54        ID054 repeat_element     FALSE
Gene55        ID055 repeat_element     FALSE
...             ...            ...       ...
Gene196       ID196 repeat_element     FALSE
Gene197       ID197 repeat_element     FALSE
Gene198       ID198 repeat_element     FALSE
Gene199       ID199 repeat_element     FALSE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(chr2, <span class="st">"counts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
Gene51      32      45     139     261     153      95
Gene52     113      80      80      61      72     131
Gene53     118      24      60     153       5     103
Gene54      20       1      35       4      89      49
Gene55     249      58      43     307      51      33
Gene56     215      96      28     198      50      81</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(chr2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 150 ranges and 3 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type      Keep
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
   Gene51     chr2 839636-839735      + |       ID051 protein_coding      TRUE
   Gene52     chr2 388487-388586      + |       ID052         lncRNA     FALSE
   Gene53     chr2 357210-357309      + |       ID053 repeat_element     FALSE
   Gene54     chr2 230565-230664      + |       ID054 repeat_element     FALSE
   Gene55     chr2 491399-491498      + |       ID055 repeat_element     FALSE
      ...      ...           ...    ... .         ...            ...       ...
  Gene196     chr2 268871-268970      + |       ID196 repeat_element     FALSE
  Gene197     chr2 599914-600013      + |       ID197 repeat_element     FALSE
  Gene198     chr2 477464-477563      + |       ID198 repeat_element     FALSE
  Gene199     chr2 461424-461523      - |       ID199 repeat_element     FALSE
  Gene200     chr2 619939-620038      + |       ID200 protein_coding      TRUE
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>There’s also a few shortcuts on range operations using GRanges/SummarizedExperiments. See the help pages for <code>%over</code>, <code>%within%</code>, and <code>%outside%</code>. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(se[se <span class="sc">%over%</span> roi, ], <span class="fu">subsetByOverlaps</span>(se, roi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</section>
<section id="combining-subsetting-operations" class="level3" data-number="2.6.4">
<h3 data-number="2.6.4" class="anchored" data-anchor-id="combining-subsetting-operations"><span class="header-section-number">2.6.4</span> Combining subsetting operations</h3>
<p>Of course you don’t have to perform one subsetting operation at a time. Like base R you can combine multiple expressions to subset a <code>SummarizedExperiment</code> object.</p>
<p>For example, to select only features labeled as repeat_elements and the Sample from ‘Batch’ A in the ‘Control’ group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a>(selected <span class="ot">&lt;-</span> se[</span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowData</span>(se)<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"repeat_element"</span>,</span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>  se<span class="sc">$</span>Treatment <span class="sc">==</span> <span class="st">"Control"</span> <span class="sc">&amp;</span></span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>    se<span class="sc">$</span>Batch <span class="sc">==</span> <span class="st">"A"</span></span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 72 1 
metadata(0):
assays(3): counts cpm logcounts
rownames(72): Gene5 Gene6 ... Gene198 Gene199
rowData names(3): feature_id gene_type Keep
colnames(1): Sample1
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
</section>
</section>
<section id="saving-a-summarizedexperiment" class="level2" data-number="2.7">
<h2 data-number="2.7" class="anchored" data-anchor-id="saving-a-summarizedexperiment"><span class="header-section-number">2.7</span> Saving a <code>SummarizedExperiment</code></h2>
<p>Since <code>SummarizedExperiments</code> keep basically all information about an experiment in one place, it is convenient to save the entire <code>SummarizedExperiment</code> object so that you can pick up an analysis where you left off or even to facilitate better sharing of data between collaborators.</p>
<p>You can save the entire <code>SummarizedExperiment</code> object with:</p>
<p><code>saveRDS(se, "/path/to/se.rds")</code></p>
<p>And when you want to read the same object back into R for your next analysis you can do so with:</p>
<p><code>se &lt;- readRDS("/path/to/se.rds")</code></p>
</section>
<section id="summarizedexperiments-in-the-real-world" class="level2" data-number="2.8">
<h2 data-number="2.8" class="anchored" data-anchor-id="summarizedexperiments-in-the-real-world"><span class="header-section-number">2.8</span> <code>SummarizedExperiments</code> in the real world</h2>
<p>If you’re working with any <a href="https://bioconductor.org/">Bioconductor</a> packages it’s likely that the object you’re working with either is a <code>SummarizedExperiment</code> or is inherited from one. For example, the <code>DESeqDataSet</code> from the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 package</a> and <code>BSseq</code> objects from the <a href="https://bioconductor.org/packages/release/bioc/html/bsseq.html">bsseq package</a> both inherit from a <code>SummarizedExperiment</code> and thus retain all of the same functionality as above. If you go to the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> landing page and click “See More” under details you can see all of the packages that depend on <code>SummarizedExperiment</code>.</p>
<p>Also, many common methods are also implemented for <code>SummarizedExperiment</code> objects. For example, to simplify calculating counts-per-million above I could have simply used the <code>edgeR::cpm()</code> directly on the <code>SummarizedExperiment</code> object. Many functions in bioconductor packages know how to deal directly with <code>SummarizedExperiments</code> so you don’t ever have to take the trouble extracting components and performing tedious calculations yourself.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I also left out any discussion of the <code>metadata()</code> slot of the <code>SummarizedExperiment</code>. The metadata slot is simply a list of any R object that contains information about the experiment. The <code>metadata</code> in the metadata slots are not subjected to the same subsetting rules as the other slots. In practice this assay contains additional information about the experiment as a whole. For example, I typically store bootstrap alignments for each sample here.</p>
<p>To add something to the <code>SummarizedExperiment</code> metadata slot you can do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(se)<span class="sc">$</span>additional_info <span class="ot">&lt;-</span> <span class="st">"Experiment performed on 6 samples with three replicates each"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And to retrieve this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(se)<span class="sc">$</span>additional_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Experiment performed on 6 samples with three replicates each"</code></pre>
</div>
</div>
</section>
<section id="closing-thoughts" class="level2" data-number="2.9">
<h2 data-number="2.9" class="anchored" data-anchor-id="closing-thoughts"><span class="header-section-number">2.9</span> Closing thoughts</h2>
<p>Hopefully this was enough information to get you started using <code>SummarizedExperiments</code>. There’s many things I left out such as different <a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html">backings</a> for storing out of memory data, a tidyverse <a href="https://stemangiola.github.io/tidySummarizedExperiment/">interface</a> to <code>SummarizedExperiment</code> objects, <a href="https://bioconductor.org/packages/release/bioc/vignettes/mia/inst/doc/mia.html">TreeSummarizedExperiments</a> for microbiome data, <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">MultiAssayExperiments</a> for dealing with experiments containing multiomics data, and much more.</p>
<p>Please let me know your thoughts and if anything needs more clarification.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./data-best-practices.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Some data best practices</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./data-viz.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Effective data visualization</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>