<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>R programming basics – 2025 Coriell Bioinformatics Internship</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./data-viz.html" rel="next">
<link href="./linux.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-128e0f949f8f902082bc899829804d79.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./r-basics.html">R programming basics</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">2025 Coriell Bioinformatics Internship</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Welcome!</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-best-practices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">‘Best’ practices for data projects</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./linux.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Command line basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-basics.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">R programming basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Effective data visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./presenting-science.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Scientific presentations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summarized-experiments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Practical introduction to SummarizedExperiments</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-expression-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-methylation-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Differential Methylation Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction-to-programming-in-r" id="toc-introduction-to-programming-in-r" class="nav-link active" data-scroll-target="#introduction-to-programming-in-r">Introduction to programming in R</a></li>
  <li><a href="#subsetting-in-r" id="toc-subsetting-in-r" class="nav-link" data-scroll-target="#subsetting-in-r">Subsetting in R</a>
  <ul class="collapse">
  <li><a href="#subsetting-vectors" id="toc-subsetting-vectors" class="nav-link" data-scroll-target="#subsetting-vectors">Subsetting vectors</a></li>
  <li><a href="#subsetting-data.frames" id="toc-subsetting-data.frames" class="nav-link" data-scroll-target="#subsetting-data.frames">Subsetting data.frames</a></li>
  <li><a href="#subsetting-lists" id="toc-subsetting-lists" class="nav-link" data-scroll-target="#subsetting-lists">Subsetting Lists</a></li>
  <li><a href="#subsetting-matrices" id="toc-subsetting-matrices" class="nav-link" data-scroll-target="#subsetting-matrices">Subsetting matrices</a></li>
  </ul></li>
  <li><a href="#data-wrangling" id="toc-data-wrangling" class="nav-link" data-scroll-target="#data-wrangling">Data Wrangling</a>
  <ul class="collapse">
  <li><a href="#data-import" id="toc-data-import" class="nav-link" data-scroll-target="#data-import">Data import</a></li>
  <li><a href="#viewing-data" id="toc-viewing-data" class="nav-link" data-scroll-target="#viewing-data">Viewing data</a></li>
  <li><a href="#splitting-data" id="toc-splitting-data" class="nav-link" data-scroll-target="#splitting-data">Splitting data</a></li>
  <li><a href="#plotting-data" id="toc-plotting-data" class="nav-link" data-scroll-target="#plotting-data">Plotting data</a></li>
  <li><a href="#filtering-data" id="toc-filtering-data" class="nav-link" data-scroll-target="#filtering-data">Filtering data</a></li>
  <li><a href="#reordering-data" id="toc-reordering-data" class="nav-link" data-scroll-target="#reordering-data">Reordering data</a></li>
  <li><a href="#selecting-columns-of-data" id="toc-selecting-columns-of-data" class="nav-link" data-scroll-target="#selecting-columns-of-data">Selecting columns of data</a></li>
  <li><a href="#modifying-columns-of-data" id="toc-modifying-columns-of-data" class="nav-link" data-scroll-target="#modifying-columns-of-data">Modifying columns of data</a></li>
  </ul></li>
  <li><a href="#string-operations" id="toc-string-operations" class="nav-link" data-scroll-target="#string-operations">String operations</a></li>
  <li><a href="#control-flow" id="toc-control-flow" class="nav-link" data-scroll-target="#control-flow">Control flow</a></li>
  <li><a href="#functions" id="toc-functions" class="nav-link" data-scroll-target="#functions">Functions</a></li>
  <li><a href="#functional-programming" id="toc-functional-programming" class="nav-link" data-scroll-target="#functional-programming">Functional programming</a></li>
  <li><a href="#resources" id="toc-resources" class="nav-link" data-scroll-target="#resources">Resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">R programming basics</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="introduction-to-programming-in-r" class="level2">
<h2 class="anchored" data-anchor-id="introduction-to-programming-in-r">Introduction to programming in R</h2>
<p>There are tons of great resources for learning R. <a href="https://r4ds.had.co.nz/">R for Data Science</a> is probably the most popular resource for new useRs to get up to speed with slicing and dicing data in R. The <em>R for Data Science</em> book, however, is taught from the perspective of the <a href="https://www.tidyverse.org/">Tidyverse</a>. The <em>Tidyverse</em> is an opinionated set of packages and functions that help users perform data manipulations primarily on data.frames. While these packages and functions can be great for experienced users by providing ergonomic and consistent interfaces for data.frame manipulation, it is my personal belief that new users should first learn the base language, especially if their goal is to perform bioinformatics analysis.</p>
<p>Bioinformatics tools rely heavily on subsetting and matrix manipulations. In my experience, users who start learning R using only function from the <em>Tidyverse</em> have a difficult time understanding matrix manipulations and subsetting operations common in bioinformatics workflows. This becomes especially important when using <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiments</a> - the backbone of many bioinformatics data structures in R.</p>
<p>For this reason, we’re going to focus on learning R from the ground up using functions that exist primarily in the base language. A great resource for learning base R quickly is Norm Matloff’s <em>fasteR</em> which can be found <a href="https://github.com/matloff/fasteR">here</a>.</p>
</section>
<section id="subsetting-in-r" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-in-r">Subsetting in R</h2>
<p>You may have only ever encountered R from the perspective of the <a href="https://www.tidyverse.org/">tidyverse</a>. <code>tidyverse</code> functions provide useful abstractions for munging <a href="https://vita.had.co.nz/papers/tidy-data.html">tidy data</a> however, most genomics data is often best represented and operated on as matrices. Keeping your data in matrix format can provide many benefits as far as speed and code clarity, which in turn helps to ensure correctness. You can think of matrices as just fancy 2D versions of vectors. So what are vectors?</p>
<p>Vectors are the main building blocks of most R analyses. Whenever you use the <code>c()</code> function (‘concatenate’), like: <code>x &lt;- c('a', 'b', 'c')</code> you’re creating a vector. Vectors hold R objects and are the building block of more complex structures in R.</p>
<p>NOTE: the following is heavily inspired by Norm Matloff’s excellent <a href="https://github.com/matloff/fasteR">fasteR</a> tutorial. Take a look there to get a brief and concise overview base R. You should also check out the first few chapters of Hadley Wickham’s amazing book <a href="https://adv-r.hadley.nz/">Advanced R</a>. The <a href="http://adv-r.had.co.nz/">first edition</a> contains some more information on base R.</p>
<section id="subsetting-vectors" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-vectors">Subsetting vectors</h3>
<p>Below, we’ll use the built-in R constant called <code>LETTERS</code>. The <code>LETTERS</code> vector is simply a ‘list’ of all uppercase letters in the Roman alphabet.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>LETTERS</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S"
[20] "T" "U" "V" "W" "X" "Y" "Z"</code></pre>
</div>
</div>
<p>We can subset the vector by position. For example, to get the 3rd letter we use the <code>[</code> operator and the position we want to extract.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C"</code></pre>
</div>
</div>
<p>We can also use a range of positions. The notation <code>3:7</code> is a shortcut that generates the numbers, 3, 4, 5, 6, 7.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="dv">3</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "C" "D" "E" "F" "G"</code></pre>
</div>
</div>
<p>We don’t have to select sequential elements either. We can extract elements by using another vector of positions.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>LETTERS[<span class="fu">c</span>(<span class="dv">7</span>, <span class="dv">5</span>, <span class="dv">14</span>, <span class="dv">14</span>, <span class="dv">1</span>, <span class="dv">18</span>, <span class="dv">15</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "G" "E" "N" "N" "A" "R" "O"</code></pre>
</div>
</div>
<p>Vectors become really powerful when we start combining them with logical operations. R supports all of the usual logical and comparison operators you can expect from a programming language, <code>&lt;</code>, <code>&gt;</code>, <code>==</code>, <code>!=</code>, <code>&lt;=</code>, <code>&gt;=</code>, <code>%in%</code>, <code>&amp;</code> and <code>|</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>my_favorite_letters <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>)</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># See that this produces a logical vector of (TRUE/FALSE) values</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co"># TRUE when LETTERS is one of my_favorite_letters and FALSE otherwise</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>LETTERS <span class="sc">%in%</span> my_favorite_letters</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[13] FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE FALSE
[25] FALSE FALSE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can use that same expression to filter the vector</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>LETTERS[LETTERS <span class="sc">%in%</span> my_favorite_letters]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "A" "B" "C"</code></pre>
</div>
</div>
<p>This same kind of subsetting works on vectors that contain numeric data as well. For example, we can filter the measurements of annual flow of water through the Nile river like so:</p>
<p><em>Nile is another built-in dataset</em></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Any values strictly greater than 1200</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>Nile[Nile <span class="sc">&gt;</span> <span class="dv">1200</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1210 1230 1370 1210 1250 1260 1220</code></pre>
</div>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Any even number - `%%` is the modulus operator</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>Nile[Nile <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">==</span> <span class="dv">0</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 1120 1160 1210 1160 1160 1230 1370 1140 1110  994 1020  960 1180  958 1140
[16] 1100 1210 1150 1250 1260 1220 1030 1100  774  840  874  694  940  916  692
[31] 1020 1050  726  456  824  702 1120 1100  832  764  768  864  862  698  744
[46]  796 1040  944  984  822 1010  676  846  812  742 1040  860  874  848  890
[61]  744  838 1050  918  986 1020  906 1170  912  746  718  714  740</code></pre>
</div>
</div>
<p>At this point it’s important to take a step back and appreciate what R is doing. Each of the comparison operators that we used above is <em>vectorized</em>. This means that the comparison is applied to all elements of the vector at one time. If you’re used to a programming language like Python this might seem foreign at first. In Python, you would have to write a list comprehension to filter observations from a list that meet a certain condition. For example, <code>[x for x in Nile if x &gt; 1200]</code>. However in R, most functions and operators are <em>vectorized</em> allowing us to do things like <code>Nile &gt; 1200</code> and have the comparison applied to all of the elements of the vector automatically.</p>
</section>
<section id="subsetting-data.frames" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-data.frames">Subsetting data.frames</h3>
<p>But these are just one dimensional vectors. In R, we usually deal with data.frames (tibbles for you tidyverse folks) and matrices. Lucky for us, the subsetting operations we learned for vectors work the same way for data.frames and matrices.</p>
<p>Let’s take a look at the built-in <code>ToothGrowth</code> dataset. The data consists of the length of odontoblasts in 60 guinea pigs receiving one of three levels of vitamin C by one of two delivery methods.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
1  4.2   VC  0.5
2 11.5   VC  0.5
3  7.3   VC  0.5
4  5.8   VC  0.5
5  6.4   VC  0.5
6 10.0   VC  0.5</code></pre>
</div>
</div>
<p>The dollar sign <code>$</code> is used to extract an individual column from the data.frame, which is just a vector.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth<span class="sc">$</span>len)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  4.2 11.5  7.3  5.8  6.4 10.0</code></pre>
</div>
</div>
<p>We can also use the <code>[[</code> to get the same thing. Double-brackets come in handy when your columns are not valid R names since <code>$</code> only works when columns are valid names.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ToothGrowth[[<span class="st">"len"</span>]])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  4.2 11.5  7.3  5.8  6.4 10.0</code></pre>
</div>
</div>
<p>When subsetting a data.frame in base R, the general scheme is:</p>
<pre><code>df[the rows you want, the columns you want]</code></pre>
<p>So in order to get the 5th row of the first column we could do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[<span class="dv">5</span>, <span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 6.4</code></pre>
</div>
</div>
<p>Again, we can combine this kind of thinking to extract rows and columns matching logical conditions. For example, if we want to get all of the animals administered orange juice (‘OJ’)</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    len supp dose
31 15.2   OJ  0.5
32 21.5   OJ  0.5
33 17.6   OJ  0.5
34  9.7   OJ  0.5
35 14.5   OJ  0.5
36 10.0   OJ  0.5
37  8.2   OJ  0.5
38  9.4   OJ  0.5
39 16.5   OJ  0.5
40  9.7   OJ  0.5
41 19.7   OJ  1.0
42 23.3   OJ  1.0
43 23.6   OJ  1.0
44 26.4   OJ  1.0
45 20.0   OJ  1.0
46 25.2   OJ  1.0
47 25.8   OJ  1.0
48 21.2   OJ  1.0
49 14.5   OJ  1.0
50 27.3   OJ  1.0
51 25.5   OJ  2.0
52 26.4   OJ  2.0
53 22.4   OJ  2.0
54 24.5   OJ  2.0
55 24.8   OJ  2.0
56 30.9   OJ  2.0
57 26.4   OJ  2.0
58 27.3   OJ  2.0
59 29.4   OJ  2.0
60 23.0   OJ  2.0</code></pre>
</div>
</div>
<p>We can also combine logical statements. For example, to get all of the rows for animals administered orange juice and with odontoblast length (‘len’) less than 10.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> ToothGrowth<span class="sc">$</span>len <span class="sc">&lt;</span> <span class="dv">10</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
34 9.7   OJ  0.5
37 8.2   OJ  0.5
38 9.4   OJ  0.5
40 9.7   OJ  0.5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="co"># We can also use the bracket notation to select rows and columns at the same time</span></span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Although this gets a little difficult to read</span></span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a>ToothGrowth[ToothGrowth<span class="sc">$</span>supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> ToothGrowth<span class="sc">$</span>len <span class="sc">&lt;</span> <span class="dv">10</span>, <span class="fu">c</span>(<span class="st">"len"</span>, <span class="st">"supp"</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp
34 9.7   OJ
37 8.2   OJ
38 9.4   OJ
40 9.7   OJ</code></pre>
</div>
</div>
<p>It gets annoying typing <code>ToothGrowth</code> every time we want to subset the data.frame. Base R has a very useful function called <code>subset()</code> that can help us type less. <code>subset()</code> essentially ‘looks inside’ the data.frame for the given columns and evaluates the expression without having to explicitly tell R where to find the columns. Think of it like <code>dplyr::filter()</code>, if you are familiar with that function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">subset</span>(ToothGrowth, supp <span class="sc">==</span> <span class="st">"OJ"</span> <span class="sc">&amp;</span> len <span class="sc">&lt;</span> <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   len supp dose
34 9.7   OJ  0.5
37 8.2   OJ  0.5
38 9.4   OJ  0.5
40 9.7   OJ  0.5</code></pre>
</div>
</div>
</section>
<section id="subsetting-lists" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-lists">Subsetting Lists</h3>
<p>Another data structure to be aware of, which is used frequently, is the <code>List</code>. We’ve actually already encountered Lists above. data.frames are really just Lists where each vector contains the same data type and all List elements are the same length.</p>
<p>We can create a List in R using the <code>list()</code> function. Notice how each list element has a name and can contain a different type of data and number of data elements</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>l <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">element1 =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="dv">12</span>, <span class="dv">3</span>, <span class="dv">6</span>, <span class="dv">12</span>, <span class="dv">13</span>, <span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">6</span>, <span class="dv">3</span>, <span class="dv">7</span>),</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">element2 =</span> <span class="fu">c</span>(<span class="st">"a"</span>, <span class="st">"b"</span>, <span class="st">"c"</span>),</span>
<span id="cb34-4"><a href="#cb34-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">element3 =</span> <span class="fu">c</span>(<span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>, <span class="cn">FALSE</span>),</span>
<span id="cb34-5"><a href="#cb34-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">element4 =</span> <span class="fu">c</span>(<span class="fl">0.001</span>, <span class="fl">0.05</span>, <span class="fl">0.86</span>, <span class="fl">1.098</span>, <span class="fl">345.0</span>)</span>
<span id="cb34-6"><a href="#cb34-6" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Lists can be tricky at first. To extract the data from a particular list element you can use the <code>[[</code> or the <code>$</code> (as in the case of data.frames above). Like vectors, you can use either the index or the name of the element you wish to extract.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>l[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1 10 12  3  6 12 13  2  5  6  3  7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>l[[<span class="st">"element1"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1 10 12  3  6 12 13  2  5  6  3  7</code></pre>
</div>
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>l<span class="sc">$</span>element1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  1 10 12  3  6 12 13  2  5  6  3  7</code></pre>
</div>
</div>
<p>What is returned if you only use the single bracket <code>[</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>l[<span class="dv">1</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$element1
 [1]  1 10 12  3  6 12 13  2  5  6  3  7</code></pre>
</div>
</div>
<p>You get another List, but now with a single element. This behavior might seem unintuitive at first, but it can be very useful for creating new lists.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>numeric_l <span class="ot">&lt;-</span> l[<span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">4</span>)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="subsetting-matrices" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-matrices">Subsetting matrices</h3>
<p>Matrices behave much like data.frames but unlike data.frames matrices can only contain one type of data. This might sound like a limitation at first but you’ll soon come to realize that matrices are very powerful (and fast) to work with in R.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Create some random data that looks like methylation values</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>(m <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">runif</span>(<span class="dv">6</span> <span class="sc">*</span> <span class="dv">10</span>),</span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ncol =</span> <span class="dv">6</span>,</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(</span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"CpG."</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>),</span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">paste0</span>(<span class="st">"Sample"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>         Sample1    Sample2   Sample3    Sample4   Sample5    Sample6
CpG.1  0.2875775 0.95683335 0.8895393 0.96302423 0.1428000 0.04583117
CpG.2  0.7883051 0.45333416 0.6928034 0.90229905 0.4145463 0.44220007
CpG.3  0.4089769 0.67757064 0.6405068 0.69070528 0.4137243 0.79892485
CpG.4  0.8830174 0.57263340 0.9942698 0.79546742 0.3688455 0.12189926
CpG.5  0.9404673 0.10292468 0.6557058 0.02461368 0.1524447 0.56094798
CpG.6  0.0455565 0.89982497 0.7085305 0.47779597 0.1388061 0.20653139
CpG.7  0.5281055 0.24608773 0.5440660 0.75845954 0.2330341 0.12753165
CpG.8  0.8924190 0.04205953 0.5941420 0.21640794 0.4659625 0.75330786
CpG.9  0.5514350 0.32792072 0.2891597 0.31818101 0.2659726 0.89504536
CpG.10 0.4566147 0.95450365 0.1471136 0.23162579 0.8578277 0.37446278</code></pre>
</div>
</div>
<p>If we want to extract the value for CpG.3 for Sample3</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">3</span>, <span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.6405068</code></pre>
</div>
</div>
<p>Or all values of CpG.3 for every sample</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>m[<span class="dv">3</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample1   Sample2   Sample3   Sample4   Sample5   Sample6 
0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or refer to the row by it's name</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>m[<span class="st">"CpG.3"</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  Sample1   Sample2   Sample3   Sample4   Sample5   Sample6 
0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248 </code></pre>
</div>
</div>
<p>Or all CpGs for Sample3</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>m[, <span class="dv">3</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    CpG.1     CpG.2     CpG.3     CpG.4     CpG.5     CpG.6     CpG.7     CpG.8 
0.8895393 0.6928034 0.6405068 0.9942698 0.6557058 0.7085305 0.5440660 0.5941420 
    CpG.9    CpG.10 
0.2891597 0.1471136 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Or refer to the column by it's name</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>m[, <span class="st">"Sample3"</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    CpG.1     CpG.2     CpG.3     CpG.4     CpG.5     CpG.6     CpG.7     CpG.8 
0.8895393 0.6928034 0.6405068 0.9942698 0.6557058 0.7085305 0.5440660 0.5941420 
    CpG.9    CpG.10 
0.2891597 0.1471136 </code></pre>
</div>
</div>
<p>We can also apply a mask to the entire matrix at once. For example, the following will mark any value that is greater than 0.5 with <code>TRUE</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>m <span class="sc">&gt;</span> <span class="fl">0.5</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
CpG.1    FALSE    TRUE    TRUE    TRUE   FALSE   FALSE
CpG.2     TRUE   FALSE    TRUE    TRUE   FALSE   FALSE
CpG.3    FALSE    TRUE    TRUE    TRUE   FALSE    TRUE
CpG.4     TRUE    TRUE    TRUE    TRUE   FALSE   FALSE
CpG.5     TRUE   FALSE    TRUE   FALSE   FALSE    TRUE
CpG.6    FALSE    TRUE    TRUE   FALSE   FALSE   FALSE
CpG.7     TRUE   FALSE    TRUE    TRUE   FALSE   FALSE
CpG.8     TRUE   FALSE    TRUE   FALSE   FALSE    TRUE
CpG.9     TRUE   FALSE   FALSE   FALSE   FALSE    TRUE
CpG.10   FALSE    TRUE   FALSE   FALSE    TRUE   FALSE</code></pre>
</div>
</div>
<p>We can use this kind of masking to filter rows of the matrix using some very helpful base R functions that operate on matrices. For example, to get only those CpGs where 3 or more samples have a value &gt; 0.5 we can use the <code>rowSums()</code> like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>m[<span class="fu">rowSums</span>(m <span class="sc">&gt;</span> <span class="fl">0.5</span>) <span class="sc">&gt;</span> <span class="dv">3</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1   Sample2   Sample3   Sample4   Sample5   Sample6
CpG.3 0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248
CpG.4 0.8830174 0.5726334 0.9942698 0.7954674 0.3688455 0.1218993</code></pre>
</div>
</div>
<p>This pattern is very common when dealing with sequencing data. Base R functions like <code>rowSums()</code> and <code>colMeans()</code> are specialized to operate over matrices and are the most efficient way to summarize matrix data. The R package <a href="https://github.com/HenrikBengtsson/matrixStats">matrixStats</a> also contains highly optimized functions for operating on matrices.</p>
<p>Compare the above to the <code>tidy</code> solution given the same matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>tidyr<span class="sc">::</span><span class="fu">as_tibble</span>(m, <span class="at">rownames =</span> <span class="st">"CpG"</span>) <span class="sc">|&gt;</span></span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="sc">!</span>CpG, <span class="at">names_to =</span> <span class="st">"SampleName"</span>, <span class="at">values_to =</span> <span class="st">"beta"</span>) <span class="sc">|&gt;</span></span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">group_by</span>(CpG) <span class="sc">|&gt;</span></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">n =</span> <span class="fu">sum</span>(beta <span class="sc">&gt;</span> <span class="fl">0.5</span>)) <span class="sc">|&gt;</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(n <span class="sc">&gt;</span> <span class="dv">3</span>) <span class="sc">|&gt;</span></span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">id_cols =</span> CpG, <span class="at">names_from =</span> <span class="st">"SampleName"</span>, <span class="at">values_from =</span> <span class="st">"beta"</span>) <span class="sc">|&gt;</span></span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>  tibble<span class="sc">::</span><span class="fu">column_to_rownames</span>(<span class="at">var =</span> <span class="st">"CpG"</span>) <span class="sc">|&gt;</span></span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">data.matrix</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1   Sample2   Sample3   Sample4   Sample5   Sample6
CpG.3 0.4089769 0.6775706 0.6405068 0.6907053 0.4137243 0.7989248
CpG.4 0.8830174 0.5726334 0.9942698 0.7954674 0.3688455 0.1218993</code></pre>
</div>
</div>
<p><em>There’s probably some kind of tidy solution using <code>across()</code> that I’m missing but this is how most of the tidy code in the wild that I have seen looks</em></p>
</section>
</section>
<section id="data-wrangling" class="level2">
<h2 class="anchored" data-anchor-id="data-wrangling">Data Wrangling</h2>
<p>Now that we’ve got a handle on some different R data types, and how to slice and dice them, we can start learning basic data cleaning and exploratory data analysis. We’ll focus on using data.frames since they’re the primary workhorse of data analysis in R. But remember, data.frames are just lists of vectors with some special rules, so many concepts you learn will apply to data.frames but also apply to vectors and lists.</p>
<section id="data-import" class="level3">
<h3 class="anchored" data-anchor-id="data-import">Data import</h3>
<p>We’ll use the built in <code>penguins_raw</code> dataset to learn some basic data cleaning. This dataset is built into R version 4.5 so you can just load it by running <code>penguins_raw</code> if you have that R version installed. However, to illustrate data import, we’ll read in the data from an external source.</p>
<p>The <code>read.csv()</code> function can read in comma-separated value files that are located either on your local machine or from remote sources if provided a URL.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a>url <span class="ot">&lt;-</span> <span class="st">"https://raw.githubusercontent.com/allisonhorst/palmerpenguins/refs/heads/main/inst/extdata/penguins_raw.csv"</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>penguins <span class="ot">&lt;-</span> <span class="fu">read.csv</span>(url)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The <code>read.csv()</code> function has many options for reading in data. If you want to learn about all of the options any particular R function has, you can prefix the function name with a <code>?</code> like, <code>?read.csv()</code> to bring up the help documentation.</p>
</section>
<section id="viewing-data" class="level3">
<h3 class="anchored" data-anchor-id="viewing-data">Viewing data</h3>
<p>The code above read the data into data.frame that we called <code>penguins</code>. We can take a look at the first few rows of the <code>penguins</code> data.frame using the <code>head()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  studyName Sample.Number                             Species Region    Island
1   PAL0708             1 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
2   PAL0708             2 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
3   PAL0708             3 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
4   PAL0708             4 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
5   PAL0708             5 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
6   PAL0708             6 Adelie Penguin (Pygoscelis adeliae) Anvers Torgersen
               Stage Individual.ID Clutch.Completion   Date.Egg
1 Adult, 1 Egg Stage          N1A1               Yes 2007-11-11
2 Adult, 1 Egg Stage          N1A2               Yes 2007-11-11
3 Adult, 1 Egg Stage          N2A1               Yes 2007-11-16
4 Adult, 1 Egg Stage          N2A2               Yes 2007-11-16
5 Adult, 1 Egg Stage          N3A1               Yes 2007-11-16
6 Adult, 1 Egg Stage          N3A2               Yes 2007-11-16
  Culmen.Length..mm. Culmen.Depth..mm. Flipper.Length..mm. Body.Mass..g.    Sex
1               39.1              18.7                 181          3750   MALE
2               39.5              17.4                 186          3800 FEMALE
3               40.3              18.0                 195          3250 FEMALE
4                 NA                NA                  NA            NA   &lt;NA&gt;
5               36.7              19.3                 193          3450 FEMALE
6               39.3              20.6                 190          3650   MALE
  Delta.15.N..o.oo. Delta.13.C..o.oo.                       Comments
1                NA                NA Not enough blood for isotopes.
2           8.94956         -24.69454                           &lt;NA&gt;
3           8.36821         -25.33302                           &lt;NA&gt;
4                NA                NA             Adult not sampled.
5           8.76651         -25.32426                           &lt;NA&gt;
6           8.66496         -25.29805                           &lt;NA&gt;</code></pre>
</div>
</div>
<p>If we want to get a general overview of the data, we can use the <code>str()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>'data.frame':   344 obs. of  17 variables:
 $ studyName          : chr  "PAL0708" "PAL0708" "PAL0708" "PAL0708" ...
 $ Sample.Number      : int  1 2 3 4 5 6 7 8 9 10 ...
 $ Species            : chr  "Adelie Penguin (Pygoscelis adeliae)" "Adelie Penguin (Pygoscelis adeliae)" "Adelie Penguin (Pygoscelis adeliae)" "Adelie Penguin (Pygoscelis adeliae)" ...
 $ Region             : chr  "Anvers" "Anvers" "Anvers" "Anvers" ...
 $ Island             : chr  "Torgersen" "Torgersen" "Torgersen" "Torgersen" ...
 $ Stage              : chr  "Adult, 1 Egg Stage" "Adult, 1 Egg Stage" "Adult, 1 Egg Stage" "Adult, 1 Egg Stage" ...
 $ Individual.ID      : chr  "N1A1" "N1A2" "N2A1" "N2A2" ...
 $ Clutch.Completion  : chr  "Yes" "Yes" "Yes" "Yes" ...
 $ Date.Egg           : chr  "2007-11-11" "2007-11-11" "2007-11-16" "2007-11-16" ...
 $ Culmen.Length..mm. : num  39.1 39.5 40.3 NA 36.7 39.3 38.9 39.2 34.1 42 ...
 $ Culmen.Depth..mm.  : num  18.7 17.4 18 NA 19.3 20.6 17.8 19.6 18.1 20.2 ...
 $ Flipper.Length..mm.: int  181 186 195 NA 193 190 181 195 193 190 ...
 $ Body.Mass..g.      : int  3750 3800 3250 NA 3450 3650 3625 4675 3475 4250 ...
 $ Sex                : chr  "MALE" "FEMALE" "FEMALE" NA ...
 $ Delta.15.N..o.oo.  : num  NA 8.95 8.37 NA 8.77 ...
 $ Delta.13.C..o.oo.  : num  NA -24.7 -25.3 NA -25.3 ...
 $ Comments           : chr  "Not enough blood for isotopes." NA NA "Adult not sampled." ...</code></pre>
</div>
</div>
<p>There are a few external packages that are also very useful for getting summaries of data.frames. <code>Hmsic::describe()</code> and <code>skimr::skim()</code> are two standouts.</p>
<p>One of the most basic ways to get an idea of the data is to summarize each variable. There are a few functions we can use to get summaries of the data. The <code>table()</code> function will count the number of occurrences of each type in a vector.</p>
<p>For example, how many observations of each species of penguin are in the dataset?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(penguins<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
      Adelie Penguin (Pygoscelis adeliae) 
                                      152 
Chinstrap penguin (Pygoscelis antarctica) 
                                       68 
        Gentoo penguin (Pygoscelis papua) 
                                      124 </code></pre>
</div>
</div>
<p>R also provides the typical summary functions that you would expect from a statistical programming language such as <code>mean()</code>, <code>median()</code>, <code>min()</code>, and <code>max()</code>, and <code>length()</code>.</p>
<p>For example, what is the mean flipper length?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(penguins<span class="sc">$</span>Flipper.Length..mm.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] NA</code></pre>
</div>
</div>
<p>On no! This returned NA but there is clearly data in this column. What happened? Missing data is commonly observed across all data domains. NAs simply represent unknown values in this context and it’s impossible to know how to take the mean of a value that’s known with a value that’s unknown. It is for this reason that many R functions have an argument called <code>na.rm=</code>. Setting <code>na.rm=TRUE</code> in these functions tells R to ignore the NA values.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">mean</span>(penguins<span class="sc">$</span>Flipper.Length..mm., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 200.9152</code></pre>
</div>
</div>
<p>Now we can see that the mean flipper length is ~200 mm across all observations in the dataset. Another useful function is <code>summary()</code>. Running <code>summary()</code> on a numeric vector returns a lot of useful information.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(penguins<span class="sc">$</span>Flipper.Length..mm., <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   Min. 1st Qu.  Median    Mean 3rd Qu.    Max.    NA's 
  172.0   190.0   197.0   200.9   213.0   231.0       2 </code></pre>
</div>
</div>
<p>Finally, if you are using and IDE like Rstudio or Positron you can run the <code>View()</code> function on your data.frame. This will bring up an interactive data viewer.</p>
</section>
<section id="splitting-data" class="level3">
<h3 class="anchored" data-anchor-id="splitting-data">Splitting data</h3>
<p>You may have noticed above that we computed the mean flipper length across all species of penguins. But do all species have the same mean? A common pattern in R is called “split-apply-combine”. This pattern means, split the data into groups you’re interested in, apply a function to each of those groups, and combine the results. One such function that performs this operation is called <code>tapply()</code>. <code>tapply()</code> will split the data by a given variable into groups and apply a function to each group.</p>
<p>For example, to find the mean flipper length for each species we could use</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(penguins<span class="sc">$</span>Flipper.Length..mm., penguins<span class="sc">$</span>Species, mean, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Adelie Penguin (Pygoscelis adeliae) 
                                 189.9536 
Chinstrap penguin (Pygoscelis antarctica) 
                                 195.8235 
        Gentoo penguin (Pygoscelis papua) 
                                 217.1870 </code></pre>
</div>
</div>
<p>The basic format of the <code>tapply()</code> function is</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">tapply</span>(<span class="st">"data to split"</span>, <span class="st">"what to split by"</span>, <span class="st">"what to compute"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Splitting can also by applied directly to data.frames using the <code>split()</code> function. Let’s say we wanted to split the penguins data.frame into one data.frame for each species. We can use the <code>split()</code> function for this purpose.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a>by_species <span class="ot">&lt;-</span> <span class="fu">split</span>(penguins, <span class="at">f =</span> penguins<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>This function returns a list of data.frames, one for each species in the original data.frame. Use <code>names(by_species)</code> to see what each list element is named. To extract the first data.frame from this list, which contains Adelie penguin data only, we can subset the list of data.frames.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>adelie <span class="ot">&lt;-</span> by_species[[<span class="dv">1</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Performing operations on lists is such a common task in R that a function exists specifically to apply functions to list elements. This function is called <code>lapply()</code>. <code>lapply()</code> takes a list and a function and applies that function to each list element. The <code>lapply()</code> function returns the results as a list. We’ll learn more about this later in the functional programming section.</p>
<p>For example, to see how many rows are in each of the data.frames in the “by_species” list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a><span class="fu">lapply</span>(by_species, nrow)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>$`Adelie Penguin (Pygoscelis adeliae)`
[1] 152

$`Chinstrap penguin (Pygoscelis antarctica)`
[1] 68

$`Gentoo penguin (Pygoscelis papua)`
[1] 124</code></pre>
</div>
</div>
</section>
<section id="plotting-data" class="level3">
<h3 class="anchored" data-anchor-id="plotting-data">Plotting data</h3>
<p>Data cleaning and data visualization go hand-in-hand. To effectively clean data, you should be examining the changes you’re making in real time. Base R actually has very powerful graphics capabilities for quickly visualizing data. Packages like <a href="https://ggplot2.tidyverse.org/">ggplot2</a> and <a href="https://lattice.r-forge.r-project.org/">lattice</a> provide powerful alternatives to base R plots. We’ll cover <code>ggplot2</code> later. For now, base R plotting can provide all we need for exploratory analyses.</p>
<p>One of the most useful plots for numeric data is a histogram. Histograms bin the data and plot how many occurrences of a particular bin are present. This plot allows you to get an idea of the numeric summary of a variable. To plot a histogram of a numeric variable we can use the <code>hist()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(penguins<span class="sc">$</span>Flipper.Length..mm.)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-37-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The histogram has a few arguments we can use to adjust the plot. Use <code>?hist()</code> to see the full list. Below, we can adjust the axes to be more informative and modify the number of bins we’re computing.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(penguins<span class="sc">$</span>Flipper.Length..mm., </span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="dv">30</span>,</span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Flipper Length of All Palmer Penguins"</span>, </span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Flipper Length (mm)"</span>,</span>
<span id="cb83-5"><a href="#cb83-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Count"</span></span>
<span id="cb83-6"><a href="#cb83-6" aria-hidden="true" tabindex="-1"></a>     )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The distribution appears to be bimodal. Is this because there are different species present in this plot? We can check by applying the <code>hist()</code> function to each of the groups using the list of data.frames from above.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Adelie penguins </span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(by_species[[<span class="dv">1</span>]]<span class="sc">$</span>Flipper.Length..mm., </span>
<span id="cb84-3"><a href="#cb84-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="dv">20</span>, </span>
<span id="cb84-4"><a href="#cb84-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Adelie"</span>, </span>
<span id="cb84-5"><a href="#cb84-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Flipper Length (mm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-39-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Chinstrap penguins</span></span>
<span id="cb85-2"><a href="#cb85-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(by_species[[<span class="dv">2</span>]]<span class="sc">$</span>Flipper.Length..mm., </span>
<span id="cb85-3"><a href="#cb85-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="dv">20</span>, </span>
<span id="cb85-4"><a href="#cb85-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Chinstrap"</span>, </span>
<span id="cb85-5"><a href="#cb85-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Flipper Length (mm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-39-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="sourceCode cell-code" id="cb86"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb86-1"><a href="#cb86-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Gentoo penguins</span></span>
<span id="cb86-2"><a href="#cb86-2" aria-hidden="true" tabindex="-1"></a><span class="fu">hist</span>(by_species[[<span class="dv">3</span>]]<span class="sc">$</span>Flipper.Length..mm., </span>
<span id="cb86-3"><a href="#cb86-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">breaks =</span> <span class="dv">20</span>, </span>
<span id="cb86-4"><a href="#cb86-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Gentoo"</span>, </span>
<span id="cb86-5"><a href="#cb86-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Flipper Length (mm)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-39-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Histograms are useful for plotting the distribution of a single numeric variable. Often, we wish to see how two variables are related. The <code>plot()</code> function provides a way to create a scatter plot of two variables against each other on the same plot. For example, to plot the culmen length vs the culmen depth:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> penguins<span class="sc">$</span>Culmen.Length..mm., </span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> penguins<span class="sc">$</span>Culmen.Depth..mm.,</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Culmen Length (mm)"</span>,</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Culmen Depth (mm)"</span>,</span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Relationship between culmen length and depth"</span></span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a>     )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-40-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Again, these data points seem to be split by the different species of penguins present in the dataset. We can color these points using an additional variable in the call to the <code>plot()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb88"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb88-1"><a href="#cb88-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(<span class="at">x =</span> penguins<span class="sc">$</span>Culmen.Length..mm., </span>
<span id="cb88-2"><a href="#cb88-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">y =</span> penguins<span class="sc">$</span>Culmen.Depth..mm.,</span>
<span id="cb88-3"><a href="#cb88-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">col =</span> <span class="fu">factor</span>(penguins<span class="sc">$</span>Species),</span>
<span id="cb88-4"><a href="#cb88-4" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"Culmen Length (mm)"</span>,</span>
<span id="cb88-5"><a href="#cb88-5" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"Culmen Depth (mm)"</span>,</span>
<span id="cb88-6"><a href="#cb88-6" aria-hidden="true" tabindex="-1"></a>     <span class="at">main =</span> <span class="st">"Relationship between culmen length and depth"</span></span>
<span id="cb88-7"><a href="#cb88-7" aria-hidden="true" tabindex="-1"></a>     )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-41-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>You may have noticed that we did something special to that vector of species. We wrapped it in the <code>factor()</code> function. In R, factors are used when we have categorical values. R uses factors to represent the different levels of each category. It may not seem important now, but factors are very useful for statistical analysis. We’ll build on this topic shortly.</p>
<p>The <code>plot()</code> function is very versatile. You can use it to create line plots as well, to show trends of variables over time. Here is a contrived example of using <code>plot()</code> to create a simple line chart.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(</span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> adelie<span class="sc">$</span>Sample.Number, </span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">y =</span> adelie<span class="sc">$</span>Body.Mass..g., </span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">type =</span> <span class="st">"l"</span>, </span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Body mass vs sample number in Adelie penguins"</span>,</span>
<span id="cb89-6"><a href="#cb89-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">xlab =</span> <span class="st">"Sample Number"</span>,</span>
<span id="cb89-7"><a href="#cb89-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">ylab =</span> <span class="st">"Body Mass (g)"</span></span>
<span id="cb89-8"><a href="#cb89-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-42-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Scatter plots are useful for displaying two numeric values against each other. However you’ll often want to compare numeric values across categorical values. One such plot for doing this is called a boxplot. Boxplots show the median and interquartile (IQR) range for a set of data. The ‘whiskers’ of the boxplot display the 1.5 x IQR of the data. We can plot a boxplot of the flipper length for each species using the <code>boxplot()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb90"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb90-1"><a href="#cb90-1" aria-hidden="true" tabindex="-1"></a><span class="fu">boxplot</span>(</span>
<span id="cb90-2"><a href="#cb90-2" aria-hidden="true" tabindex="-1"></a>  Flipper.Length..mm. <span class="sc">~</span> Species, </span>
<span id="cb90-3"><a href="#cb90-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> penguins, </span>
<span id="cb90-4"><a href="#cb90-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Flipper length by species"</span></span>
<span id="cb90-5"><a href="#cb90-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-43-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>The <code>boxplot()</code> function is a little funny because it uses formula notation. Many functions in R accept formula notation as input. In this case, the formula notation means “plot the flipper length as a function of the species type”. You may have also noticed that we gave this function the penguins data.frame as the value for the <code>data=</code> argument. This allowed the boxplot function to use the column names without us having to explicitly say that they came from the penguins data.frame.</p>
<p>If there’s not too much data, boxplots can actually hide a lot of information. In this case, another option is to use a stripchart. A stripchart shows every data point on the plot instead of a depiction of the distribution as in the boxplot.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">stripchart</span>(</span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  Flipper.Length..mm. <span class="sc">~</span> Species, </span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> penguins, </span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">method =</span> <span class="st">"jitter"</span>,</span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pch =</span> <span class="dv">1</span>,</span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">main =</span> <span class="st">"Flipper length by species"</span>,</span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">vertical =</span> <span class="cn">TRUE</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>Bar charts are another common way that people show summaries of data. To display a barplot of data, we can use the <code>barplot()</code> function. To create a barplot of the mean flipper length for each species we first need to summarize the flipper length for each group and then supply these summaries to the <code>barplot()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1" aria-hidden="true" tabindex="-1"></a>species_means <span class="ot">&lt;-</span> <span class="fu">tapply</span>(penguins<span class="sc">$</span>Flipper.Length..mm., penguins<span class="sc">$</span>Species, mean, <span class="at">na.rm =</span> <span class="cn">TRUE</span>)</span>
<span id="cb92-2"><a href="#cb92-2" aria-hidden="true" tabindex="-1"></a><span class="fu">barplot</span>(species_means, <span class="at">main =</span> <span class="st">"Mean flipper length by species"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-45-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>There are many ways to modify these default plot types and make the charts look great. Take the time to look into some of base R graphing capabilities. A really good tutorial for learning about base R plotting can be found <a href="http://karolis.koncevicius.lt/posts/r_base_plotting_without_wrappers/">here</a>. Also take a look at <code>?pairs()</code>, <code>?dotchart()</code>, and <code>?coplot()</code> for some other useful base R graphics functions for creating quick plots.</p>
</section>
<section id="filtering-data" class="level3">
<h3 class="anchored" data-anchor-id="filtering-data">Filtering data</h3>
<p>We’ve already learned how to subset a data.frame but let’s just take a step back and quickly revisit subsetting data.frames. To subset a data.frame means to select rows of a data.frame based on some condition that you’re interested in. This subsetting can be accomplished using the <code>[</code> operator and setting conditions by specifying <code>df[the rows we want, the cols we want]</code> syntax.</p>
<p>For example, in the above plots it looked like the Gentoo penguins typically had the largest flippers. It looked like a value &gt; 205 roughly separated the Gentoo penguins from the others. How could we select the rows where the flipper length is greater than or equal to 205 and count the species types?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>big_flippers <span class="ot">&lt;-</span> penguins[penguins<span class="sc">$</span>Flipper.Length..mm. <span class="sc">&gt;=</span> <span class="dv">205</span>, ]</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(big_flippers<span class="sc">$</span>Species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
      Adelie Penguin (Pygoscelis adeliae) 
                                        3 
Chinstrap penguin (Pygoscelis antarctica) 
                                        8 
        Gentoo penguin (Pygoscelis papua) 
                                      122 </code></pre>
</div>
</div>
<p>We can combine conditions as well to select on multiple conditions. For example, let’s select the female penguins with flippers &gt;= 205 mm.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins[penguins<span class="sc">$</span>Sex <span class="sc">==</span> <span class="st">"FEMALE"</span> <span class="sc">&amp;</span> penguins<span class="sc">$</span>Flipper.Length..mm. <span class="sc">&gt;=</span> <span class="dv">205</span>, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>    studyName Sample.Number                           Species Region Island
NA       &lt;NA&gt;            NA                              &lt;NA&gt;   &lt;NA&gt;   &lt;NA&gt;
153   PAL0708             1 Gentoo penguin (Pygoscelis papua) Anvers Biscoe
155   PAL0708             3 Gentoo penguin (Pygoscelis papua) Anvers Biscoe
158   PAL0708             6 Gentoo penguin (Pygoscelis papua) Anvers Biscoe
159   PAL0708             7 Gentoo penguin (Pygoscelis papua) Anvers Biscoe
161   PAL0708             9 Gentoo penguin (Pygoscelis papua) Anvers Biscoe
                 Stage Individual.ID Clutch.Completion   Date.Egg
NA                &lt;NA&gt;          &lt;NA&gt;              &lt;NA&gt;       &lt;NA&gt;
153 Adult, 1 Egg Stage         N31A1               Yes 2007-11-27
155 Adult, 1 Egg Stage         N32A1               Yes 2007-11-27
158 Adult, 1 Egg Stage         N33A2               Yes 2007-11-18
159 Adult, 1 Egg Stage         N34A1               Yes 2007-11-27
161 Adult, 1 Egg Stage         N35A1               Yes 2007-11-27
    Culmen.Length..mm. Culmen.Depth..mm. Flipper.Length..mm. Body.Mass..g.
NA                  NA                NA                  NA            NA
153               46.1              13.2                 211          4500
155               48.7              14.1                 210          4450
158               46.5              13.5                 210          4550
159               45.4              14.6                 211          4800
161               43.3              13.4                 209          4400
       Sex Delta.15.N..o.oo. Delta.13.C..o.oo. Comments
NA    &lt;NA&gt;                NA                NA     &lt;NA&gt;
153 FEMALE           7.99300         -25.51390     &lt;NA&gt;
155 FEMALE           8.14705         -25.46172     &lt;NA&gt;
158 FEMALE           7.99530         -25.32829     &lt;NA&gt;
159 FEMALE           8.24515         -25.46782     &lt;NA&gt;
161 FEMALE           8.13643         -25.32176     &lt;NA&gt;</code></pre>
</div>
</div>
<p>Another important aspect of base R data.frames is that they can utilize <code>rownames()</code>. Using <code>rownames()</code> is uncommon in workflows from the <code>tidyverse</code> or even <code>data.table</code>. However, <code>rownames()</code> can be very useful when dealing with bioinformatics data. One R package that utilizes <code>rownames()</code> extensively is the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment</a>. <code>SummarizedExperiment</code>s use rownames and colnames to ensure that data stays coordinated during an analysis.</p>
<p>You can view and set the <code>rownames()</code> on a data.frame using the <code>rownames()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View the current rownames</span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(penguins))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "1" "2" "3" "4" "5" "6"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the rownames based on a new value - we'll go over what paste() does shortly</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rownames</span>(penguins) <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="st">"row_number"</span>, <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(penguins), <span class="at">sep =</span> <span class="st">"."</span>)</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># View the new rownames</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">rownames</span>(penguins))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "row_number.1" "row_number.2" "row_number.3" "row_number.4" "row_number.5"
[6] "row_number.6"</code></pre>
</div>
</div>
<p>Above, we filtered by a column in the data.frame. We can also filter by selecting rownames.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb101"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb101-1"><a href="#cb101-1" aria-hidden="true" tabindex="-1"></a>penguins[<span class="fu">c</span>(<span class="st">"row_number.1"</span>, <span class="st">"row_number.5"</span>), ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             studyName Sample.Number                             Species Region
row_number.1   PAL0708             1 Adelie Penguin (Pygoscelis adeliae) Anvers
row_number.5   PAL0708             5 Adelie Penguin (Pygoscelis adeliae) Anvers
                Island              Stage Individual.ID Clutch.Completion
row_number.1 Torgersen Adult, 1 Egg Stage          N1A1               Yes
row_number.5 Torgersen Adult, 1 Egg Stage          N3A1               Yes
               Date.Egg Culmen.Length..mm. Culmen.Depth..mm.
row_number.1 2007-11-11               39.1              18.7
row_number.5 2007-11-16               36.7              19.3
             Flipper.Length..mm. Body.Mass..g.    Sex Delta.15.N..o.oo.
row_number.1                 181          3750   MALE                NA
row_number.5                 193          3450 FEMALE           8.76651
             Delta.13.C..o.oo.                       Comments
row_number.1                NA Not enough blood for isotopes.
row_number.5         -25.32426                           &lt;NA&gt;</code></pre>
</div>
</div>
</section>
<section id="reordering-data" class="level3">
<h3 class="anchored" data-anchor-id="reordering-data">Reordering data</h3>
<p>We can also change the order of the rows in a data.frame. This is similar to subsetting but instead rearranges the data based on some condition. To reorder the data.frame we can use the <code>order()</code> function.</p>
<p>For example, we can order the data by decreasing flipper length</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb103"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb103-1"><a href="#cb103-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins[<span class="fu">order</span>(penguins<span class="sc">$</span>Flipper.Length..mm., <span class="at">decreasing =</span> <span class="cn">TRUE</span>), ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               studyName Sample.Number                           Species Region
row_number.216   PAL0809            64 Gentoo penguin (Pygoscelis papua) Anvers
row_number.154   PAL0708             2 Gentoo penguin (Pygoscelis papua) Anvers
row_number.186   PAL0708            34 Gentoo penguin (Pygoscelis papua) Anvers
row_number.218   PAL0809            66 Gentoo penguin (Pygoscelis papua) Anvers
row_number.228   PAL0809            76 Gentoo penguin (Pygoscelis papua) Anvers
row_number.242   PAL0910            90 Gentoo penguin (Pygoscelis papua) Anvers
               Island              Stage Individual.ID Clutch.Completion
row_number.216 Biscoe Adult, 1 Egg Stage         N19A2               Yes
row_number.154 Biscoe Adult, 1 Egg Stage         N31A2               Yes
row_number.186 Biscoe Adult, 1 Egg Stage         N56A2               Yes
row_number.218 Biscoe Adult, 1 Egg Stage         N20A2               Yes
row_number.228 Biscoe Adult, 1 Egg Stage         N56A2               Yes
row_number.242 Biscoe Adult, 1 Egg Stage         N14A2               Yes
                 Date.Egg Culmen.Length..mm. Culmen.Depth..mm.
row_number.216 2008-11-13               54.3              15.7
row_number.154 2007-11-27               50.0              16.3
row_number.186 2007-12-03               59.6              17.0
row_number.218 2008-11-04               49.8              16.8
row_number.228 2008-11-06               48.6              16.0
row_number.242 2009-11-25               52.1              17.0
               Flipper.Length..mm. Body.Mass..g.  Sex Delta.15.N..o.oo.
row_number.216                 231          5650 MALE           8.49662
row_number.154                 230          5700 MALE           8.14756
row_number.186                 230          6050 MALE           7.76843
row_number.218                 230          5700 MALE           8.47067
row_number.228                 230          5800 MALE           8.59640
row_number.242                 230          5550 MALE           8.27595
               Delta.13.C..o.oo. Comments
row_number.216         -26.84166     &lt;NA&gt;
row_number.154         -25.39369     &lt;NA&gt;
row_number.186         -25.68210     &lt;NA&gt;
row_number.218         -26.69166     &lt;NA&gt;
row_number.228         -26.71199     &lt;NA&gt;
row_number.242         -26.11657     &lt;NA&gt;</code></pre>
</div>
</div>
<p>R also provides the <code>sort()</code> function. See if you can understand the difference between <code>order()</code> and <code>sort()</code>.</p>
<p>Similar to filtering above, we can also reorder by the <code>rownames()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins[<span class="fu">order</span>(<span class="fu">rownames</span>(penguins)), ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               studyName Sample.Number                             Species
row_number.1     PAL0708             1 Adelie Penguin (Pygoscelis adeliae)
row_number.10    PAL0708            10 Adelie Penguin (Pygoscelis adeliae)
row_number.100   PAL0809           100 Adelie Penguin (Pygoscelis adeliae)
row_number.101   PAL0910           101 Adelie Penguin (Pygoscelis adeliae)
row_number.102   PAL0910           102 Adelie Penguin (Pygoscelis adeliae)
row_number.103   PAL0910           103 Adelie Penguin (Pygoscelis adeliae)
               Region    Island              Stage Individual.ID
row_number.1   Anvers Torgersen Adult, 1 Egg Stage          N1A1
row_number.10  Anvers Torgersen Adult, 1 Egg Stage          N5A2
row_number.100 Anvers     Dream Adult, 1 Egg Stage         N50A2
row_number.101 Anvers    Biscoe Adult, 1 Egg Stage         N47A1
row_number.102 Anvers    Biscoe Adult, 1 Egg Stage         N47A2
row_number.103 Anvers    Biscoe Adult, 1 Egg Stage         N49A1
               Clutch.Completion   Date.Egg Culmen.Length..mm.
row_number.1                 Yes 2007-11-11               39.1
row_number.10                Yes 2007-11-09               42.0
row_number.100               Yes 2008-11-10               43.2
row_number.101               Yes 2009-11-09               35.0
row_number.102               Yes 2009-11-09               41.0
row_number.103               Yes 2009-11-15               37.7
               Culmen.Depth..mm. Flipper.Length..mm. Body.Mass..g.    Sex
row_number.1                18.7                 181          3750   MALE
row_number.10               20.2                 190          4250   &lt;NA&gt;
row_number.100              18.5                 192          4100   MALE
row_number.101              17.9                 192          3725 FEMALE
row_number.102              20.0                 203          4725   MALE
row_number.103              16.0                 183          3075 FEMALE
               Delta.15.N..o.oo. Delta.13.C..o.oo.
row_number.1                  NA                NA
row_number.10            9.13362         -25.09368
row_number.100           8.97025         -26.03679
row_number.101           8.84451         -26.28055
row_number.102           9.01079         -26.38085
row_number.103           9.21510         -26.22530
                                           Comments
row_number.1         Not enough blood for isotopes.
row_number.10  No blood sample obtained for sexing.
row_number.100                                 &lt;NA&gt;
row_number.101                                 &lt;NA&gt;
row_number.102                                 &lt;NA&gt;
row_number.103                                 &lt;NA&gt;</code></pre>
</div>
</div>
<p>Can you figure out why the data.frame was sorted this way?</p>
</section>
<section id="selecting-columns-of-data" class="level3">
<h3 class="anchored" data-anchor-id="selecting-columns-of-data">Selecting columns of data</h3>
<p>Similarly to how we were able to select rows of data from the data.frame, we can also select columns of the data.frame. First, to see what columns are in our data.frame we can use the <code>colnames()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(penguins)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "studyName"           "Sample.Number"       "Species"            
 [4] "Region"              "Island"              "Stage"              
 [7] "Individual.ID"       "Clutch.Completion"   "Date.Egg"           
[10] "Culmen.Length..mm."  "Culmen.Depth..mm."   "Flipper.Length..mm."
[13] "Body.Mass..g."       "Sex"                 "Delta.15.N..o.oo."  
[16] "Delta.13.C..o.oo."   "Comments"           </code></pre>
</div>
</div>
<p>To select columns, we can either specify the columns we want by name</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins[, <span class="fu">c</span>(<span class="st">"studyName"</span>, <span class="st">"Species"</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             studyName                             Species
row_number.1   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.2   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.3   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.4   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.5   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.6   PAL0708 Adelie Penguin (Pygoscelis adeliae)</code></pre>
</div>
</div>
<p>Or by the column index</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(penguins[, <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">3</span>)])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             studyName                             Species
row_number.1   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.2   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.3   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.4   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.5   PAL0708 Adelie Penguin (Pygoscelis adeliae)
row_number.6   PAL0708 Adelie Penguin (Pygoscelis adeliae)</code></pre>
</div>
</div>
<p>If we want to rename a column, we can use the <code>colnames()</code> function to reassign the column name with a new name. First, we can get the index of the column we want to change. Then we can reassign that column with the new name.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb113"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb113-1"><a href="#cb113-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This returns the index of the colnames() that matches "Species"</span></span>
<span id="cb113-2"><a href="#cb113-2" aria-hidden="true" tabindex="-1"></a>which_col_to_change <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">colnames</span>(penguins) <span class="sc">==</span> <span class="st">"Species"</span>)</span>
<span id="cb113-3"><a href="#cb113-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb113-4"><a href="#cb113-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename "Species" to "species"</span></span>
<span id="cb113-5"><a href="#cb113-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(penguins)[which_col_to_change] <span class="ot">&lt;-</span> <span class="st">"species"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>One tricky aspect of base R subsetting on data.frames has to do with selecting a single column of data. If you select a single column of data from a data.frame what you get back is a vector. To ensure that you return a data.frame when selecting a single column, add the <code>drop=FALSE</code> argument when selecting a single column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns a vector</span></span>
<span id="cb114-2"><a href="#cb114-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(penguins[, <span class="st">"species"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "character"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Returns a data.frame</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(penguins[, <span class="st">"species"</span>, <span class="at">drop =</span> <span class="cn">FALSE</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "data.frame"</code></pre>
</div>
</div>
</section>
<section id="modifying-columns-of-data" class="level3">
<h3 class="anchored" data-anchor-id="modifying-columns-of-data">Modifying columns of data</h3>
<p>Adding or removing columns of data to a data.frame is as simple as specifying the name of the column you want to add along with the data that you want to add.</p>
<p>For example, if I wanted to convert the flipper length from mm to m I could multiply each value of flipper length by 0.001.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>flipper_length_m <span class="ot">&lt;-</span> penguins<span class="sc">$</span>Flipper.Length..mm. <span class="sc">*</span> <span class="fl">0.001</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The above data transformation works because of R’s <em>vectorization</em> rule described above. What if I wanted to define a new variable based on a logical condition? For example, what if I wanted to create a new column based on whether or not a penguin was and Adelie or any other kind? A useful function for performing this action is <code>ifelse()</code>. <code>ifelse()</code> evaluates a logical condition and returns a value based on whether the condition evaluates to TRUE or FALSE.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb119"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb119-1"><a href="#cb119-1" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>is_adelie <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(</span>
<span id="cb119-2"><a href="#cb119-2" aria-hidden="true" tabindex="-1"></a>  penguins<span class="sc">$</span>species <span class="sc">==</span> <span class="st">"Adelie Penguin (Pygoscelis adeliae)"</span>, </span>
<span id="cb119-3"><a href="#cb119-3" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Adelie"</span>, </span>
<span id="cb119-4"><a href="#cb119-4" aria-hidden="true" tabindex="-1"></a>  <span class="st">"Other"</span></span>
<span id="cb119-5"><a href="#cb119-5" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>If I want to remove a column from the data.frame I can set the value of the column to <code>NULL</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>Flipper.Length..mm. <span class="ot">&lt;-</span> <span class="cn">NULL</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="string-operations" class="level2">
<h2 class="anchored" data-anchor-id="string-operations">String operations</h2>
<p>String operations are an important part of data cleaning. Base R supports many functions for transforming strings. Again, these string functions are typically <em>vectorized</em> meaning that they can easily be used to modify columns of data.frames.</p>
<p>The “species” column of the penguins data.frame is annoying to work with. We can simplify this column by creating simpler names for each species by excluding the scientific name and the word ‘penguin’ (we know they’re all penguins…)</p>
<p>One base R function we can use to find and replace text is called <code>gsub()</code>. <code>gsub()</code> takes what is called a <a href="https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-language-quick-reference">regular expression</a> to find text inside of a string and then replaces the text it finds with the text you supply. Regular expressions can become incredibly complex. With this complexity come a lot of power. It would be impossible to teach regular expression in this tutorial. Taking some time to understand the basics of regular expressions can go a long way during data cleaning.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb121"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb121-1"><a href="#cb121-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace a space followed by the word 'penguins' and any other character</span></span>
<span id="cb121-2"><a href="#cb121-2" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>species_clean <span class="ot">&lt;-</span> <span class="fu">gsub</span>(</span>
<span id="cb121-3"><a href="#cb121-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">pattern =</span> <span class="st">" penguin.*"</span>, </span>
<span id="cb121-4"><a href="#cb121-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">replacement =</span> <span class="st">""</span>, </span>
<span id="cb121-5"><a href="#cb121-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> penguins<span class="sc">$</span>species, </span>
<span id="cb121-6"><a href="#cb121-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">ignore.case =</span> <span class="cn">TRUE</span></span>
<span id="cb121-7"><a href="#cb121-7" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb121-8"><a href="#cb121-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb121-9"><a href="#cb121-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the counts for these new categories</span></span>
<span id="cb121-10"><a href="#cb121-10" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(penguins<span class="sc">$</span>species_clean)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
   Adelie Chinstrap    Gentoo 
      152        68       124 </code></pre>
</div>
</div>
<p>Another base R function that is useful for extracting text is the <code>substr()</code> function. <code>substr()</code> extracts sub-strings from the supplied text based on the index of the text. For example, to create an even simpler representation of species we could extract only the first letter from the “species” column.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb123"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb123-1"><a href="#cb123-1" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>species_simple <span class="ot">&lt;-</span> <span class="fu">substr</span>(penguins<span class="sc">$</span>species, <span class="at">start =</span> <span class="dv">1</span>, <span class="at">stop =</span> <span class="dv">1</span>)</span>
<span id="cb123-2"><a href="#cb123-2" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(penguins<span class="sc">$</span>species_simple)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
  A   C   G 
152  68 124 </code></pre>
</div>
</div>
<p>A slightly more complicated string operation is extracting text from a string. This involves two functions; one to find a match in the text and another to extract it. Let’s say we wanted to extract the word ‘penguin’ from each of the original species strings. We can do this with a combination of <code>regmatches()</code> and <code>regexpr()</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb125"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb125-1"><a href="#cb125-1" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>only_penguin <span class="ot">&lt;-</span> <span class="fu">regmatches</span>(</span>
<span id="cb125-2"><a href="#cb125-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> penguins<span class="sc">$</span>species, </span>
<span id="cb125-3"><a href="#cb125-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">m =</span> <span class="fu">regexpr</span>(<span class="at">pattern =</span> <span class="st">"penguin"</span>, </span>
<span id="cb125-4"><a href="#cb125-4" aria-hidden="true" tabindex="-1"></a>              <span class="at">text =</span> penguins<span class="sc">$</span>species, </span>
<span id="cb125-5"><a href="#cb125-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">ignore.case =</span> <span class="cn">TRUE</span>)</span>
<span id="cb125-6"><a href="#cb125-6" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>R also supports basic operations on strings like making all characters upper or lowercase. For example, to ensure <code>only_penguin</code> contains ‘penguin’ (and not ‘Penguin’) we can convert the string to lowercase</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb126"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb126-1"><a href="#cb126-1" aria-hidden="true" tabindex="-1"></a><span class="co"># lower case</span></span>
<span id="cb126-2"><a href="#cb126-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">tolower</span>(penguins<span class="sc">$</span>only_penguin))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "penguin" "penguin" "penguin" "penguin" "penguin" "penguin"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb128"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb128-1"><a href="#cb128-1" aria-hidden="true" tabindex="-1"></a><span class="co"># or upper case</span></span>
<span id="cb128-2"><a href="#cb128-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">toupper</span>(penguins<span class="sc">$</span>only_penguin))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "PENGUIN" "PENGUIN" "PENGUIN" "PENGUIN" "PENGUIN" "PENGUIN"</code></pre>
</div>
</div>
<p>Another slightly complicated string operation is string splitting. String splitting in base R can be accomplished using the <code>strsplit()</code> function. This function is a little tricky because it returns a list. Here, we split the “species” column on every space character. Using the results from the <code>strsplit()</code> function most efficiently involves some advanced R skills.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb130"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb130-1"><a href="#cb130-1" aria-hidden="true" tabindex="-1"></a>split_species <span class="ot">&lt;-</span> <span class="fu">strsplit</span>(</span>
<span id="cb130-2"><a href="#cb130-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">x =</span> penguins<span class="sc">$</span>species, </span>
<span id="cb130-3"><a href="#cb130-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">split =</span> <span class="st">" "</span></span>
<span id="cb130-4"><a href="#cb130-4" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb130-5"><a href="#cb130-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb130-6"><a href="#cb130-6" aria-hidden="true" tabindex="-1"></a><span class="co"># strsplit returns a list of character vectors split on every space</span></span>
<span id="cb130-7"><a href="#cb130-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(split_species)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[[1]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   

[[2]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   

[[3]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   

[[4]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   

[[5]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   

[[6]]
[1] "Adelie"      "Penguin"     "(Pygoscelis" "adeliae)"   </code></pre>
</div>
<div class="sourceCode cell-code" id="cb132"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb132-1"><a href="#cb132-1" aria-hidden="true" tabindex="-1"></a><span class="co"># To extract the species from this list, we can use a fancy trick</span></span>
<span id="cb132-2"><a href="#cb132-2" aria-hidden="true" tabindex="-1"></a><span class="co"># don't worry about what's happening here right now</span></span>
<span id="cb132-3"><a href="#cb132-3" aria-hidden="true" tabindex="-1"></a>penguins<span class="sc">$</span>extracted_species <span class="ot">&lt;-</span> <span class="fu">sapply</span>(split_species, <span class="st">`</span><span class="at">[[</span><span class="st">`</span>, <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="control-flow" class="level2">
<h2 class="anchored" data-anchor-id="control-flow">Control flow</h2>
<p>R contains all of the typical control flow you expect from a programming language.</p>
<p>For-loops are somewhat under-utilized by many R users but can often be the most clear for new (and experienced) users. one common for-loop idion in R is to use a for-loop with the <code>seq_along()</code> function. <code>seq_along()</code> generates an index of the given vector that you can use to loop over.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="fu">seq_along</span>(x)) {</span>
<span id="cb133-2"><a href="#cb133-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">doSomething</span>(x[i])</span>
<span id="cb133-3"><a href="#cb133-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Other looping operations in base R include <code>while</code> and <code>repeat</code>.</p>
<p>Conditional statements can be tested using if statements. These follow a familiar syntax</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb134"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb134-1"><a href="#cb134-1" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="dv">10</span></span>
<span id="cb134-2"><a href="#cb134-2" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (x <span class="sc">==</span> <span class="dv">10</span>) {</span>
<span id="cb134-3"><a href="#cb134-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"the value of x is 10"</span>)</span>
<span id="cb134-4"><a href="#cb134-4" aria-hidden="true" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb134-5"><a href="#cb134-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(<span class="st">"the value of x is something else"</span>)</span>
<span id="cb134-6"><a href="#cb134-6" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "the value of x is 10"</code></pre>
</div>
</div>
<p>If statements in R are <em>not</em> vectorized. Use <code>ifelse()</code> to perform a vectorized if statement as demonstrated above.</p>
<p>If statements are often combined with for-loops.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb136"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb136-1"><a href="#cb136-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Make some space to save the result</span></span>
<span id="cb136-2"><a href="#cb136-2" aria-hidden="true" tabindex="-1"></a>result <span class="ot">&lt;-</span> <span class="fu">vector</span>(<span class="st">"character"</span>, <span class="at">length =</span> <span class="dv">10</span>)</span>
<span id="cb136-3"><a href="#cb136-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-4"><a href="#cb136-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine if the number is even or odd</span></span>
<span id="cb136-5"><a href="#cb136-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>) {</span>
<span id="cb136-6"><a href="#cb136-6" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (i <span class="sc">%%</span> <span class="dv">2</span> <span class="sc">==</span> <span class="dv">0</span>) {</span>
<span id="cb136-7"><a href="#cb136-7" aria-hidden="true" tabindex="-1"></a>    result[i] <span class="ot">&lt;-</span> <span class="st">"even"</span></span>
<span id="cb136-8"><a href="#cb136-8" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb136-9"><a href="#cb136-9" aria-hidden="true" tabindex="-1"></a>    result[i] <span class="ot">&lt;-</span> <span class="st">"odd"</span></span>
<span id="cb136-10"><a href="#cb136-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb136-11"><a href="#cb136-11" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb136-12"><a href="#cb136-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb136-13"><a href="#cb136-13" aria-hidden="true" tabindex="-1"></a>result</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "odd"  "even" "odd"  "even" "odd"  "even" "odd"  "even" "odd"  "even"</code></pre>
</div>
</div>
<p>R also contains a <code>switch()</code> statement which can be very useful when programming new functions.</p>
</section>
<section id="functions" class="level2">
<h2 class="anchored" data-anchor-id="functions">Functions</h2>
<p>Eventually you’ll want to create your own functions to do stuff. Functions can be defined in R using the following syntax</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb138"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb138-1"><a href="#cb138-1" aria-hidden="true" tabindex="-1"></a>f <span class="ot">&lt;-</span> <span class="cf">function</span>(args) {</span>
<span id="cb138-2"><a href="#cb138-2" aria-hidden="true" tabindex="-1"></a>  expr</span>
<span id="cb138-3"><a href="#cb138-3" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>To make this a little more concrete, let’s define a function that classifies a penguin based on it’s flipper size. Our function will take in a flipper size and a threshold value for determining whether or not the penguin is “large” or “small” based on this flipper size. The function will return the resulting size designation.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb139"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb139-1"><a href="#cb139-1" aria-hidden="true" tabindex="-1"></a>determine_size <span class="ot">&lt;-</span> <span class="cf">function</span>(flipper_size, <span class="at">threshold =</span> <span class="fl">0.2</span>) {</span>
<span id="cb139-2"><a href="#cb139-2" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (flipper_size <span class="sc">&gt;=</span> threshold) {</span>
<span id="cb139-3"><a href="#cb139-3" aria-hidden="true" tabindex="-1"></a>    size_category <span class="ot">&lt;-</span> <span class="st">"large"</span></span>
<span id="cb139-4"><a href="#cb139-4" aria-hidden="true" tabindex="-1"></a>  } <span class="cf">else</span> {</span>
<span id="cb139-5"><a href="#cb139-5" aria-hidden="true" tabindex="-1"></a>    size_category <span class="ot">&lt;-</span> <span class="st">"small"</span></span>
<span id="cb139-6"><a href="#cb139-6" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb139-7"><a href="#cb139-7" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb139-8"><a href="#cb139-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(size_category)</span>
<span id="cb139-9"><a href="#cb139-9" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb139-10"><a href="#cb139-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb139-11"><a href="#cb139-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Is a 0.5 meter flipper length large or small?</span></span>
<span id="cb139-12"><a href="#cb139-12" aria-hidden="true" tabindex="-1"></a><span class="fu">determine_size</span>(<span class="at">flipper_size =</span> <span class="fl">0.5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "large"</code></pre>
</div>
</div>
</section>
<section id="functional-programming" class="level2">
<h2 class="anchored" data-anchor-id="functional-programming">Functional programming</h2>
<p>For-loops in R are very simple to understand but most R programmers use functions in the <code>apply()</code> family. These functions supply abstractions over common for-loops that make applying functions over lists much easier in R.</p>
<p>The most commonly used functional programming paradigm in R is the use of the <code>lapply()</code> function. The <code>lapply()</code> function takes a list as input, applies a function to each element of that list, and returns the results as a list. For example, above we split the penguins data.frame into a list of smaller data.frames for each species. Instead of repeating the code to create a histogram from each data.frame, let’s define a function that plots a histogram using a data.frame and then apply that function to every data.frame in our “by_species” list of data.frames</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Define a function for plotting a histogram of flipper lengths</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a>plot_histogram <span class="ot">&lt;-</span> <span class="cf">function</span>(df) {</span>
<span id="cb141-3"><a href="#cb141-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">hist</span>(df[, <span class="st">"Flipper.Length..mm."</span>], <span class="at">breaks=</span><span class="dv">20</span>)</span>
<span id="cb141-4"><a href="#cb141-4" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb141-5"><a href="#cb141-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb141-6"><a href="#cb141-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Apply this function to every data.frame</span></span>
<span id="cb141-7"><a href="#cb141-7" aria-hidden="true" tabindex="-1"></a><span class="co"># invisible() is used to hide the output from the hist() function</span></span>
<span id="cb141-8"><a href="#cb141-8" aria-hidden="true" tabindex="-1"></a><span class="fu">invisible</span>(</span>
<span id="cb141-9"><a href="#cb141-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(<span class="at">X =</span> by_species, <span class="at">FUN =</span> plot_histogram)</span>
<span id="cb141-10"><a href="#cb141-10" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-70-2.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="r-basics_files/figure-html/unnamed-chunk-70-3.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<p>R includes some other specialized functions for functional programming. Be sure to explore <code>apply()</code> for computing over rows/columns of matrices, <code>sapply()</code>and <code>vapply()</code> for applying functions over lists and returning vectors, <code>mapply()</code> for supplying multiple arguments to a list.</p>
</section>
<section id="resources" class="level2">
<h2 class="anchored" data-anchor-id="resources">Resources</h2>
<p>This only scratches the surface of programming in R. Check out these resources for more information.</p>
<ul>
<li><a href="https://github.com/matloff/fasteR">fasteR</a></li>
<li><a href="https://r4ds.had.co.nz/">R for Data Science</a></li>
<li><a href="https://adv-r.hadley.nz/">Advanced R 2nd Edition</a></li>
<li><a href="http://adv-r.had.co.nz/">Advanced R 1st Edition</a></li>
<li><a href="https://www.burns-stat.com/pages/Tutor/R_inferno.pdf">The R Inferno</a></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./linux.html" class="pagination-link" aria-label="Command line basics">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Command line basics</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./data-viz.html" class="pagination-link" aria-label="Effective data visualizations">
        <span class="nav-page-text">Effective data visualizations</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>