<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.31">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Practical introduction to SummarizedExperiments – 2025 Coriell Bioinformatics Internship</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./differential-expression-analysis.html" rel="next">
<link href="./presenting-science.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js" type="module"></script>
<script src="site_libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-e1a5c8363afafaef2c763b6775fbf3ca.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap-128e0f949f8f902082bc899829804d79.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating quarto-light">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" role="button" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./summarized-experiments.html">Practical introduction to SummarizedExperiments</a></li></ol></nav>
        <a class="flex-grow-1" role="navigation" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
      <button type="button" class="btn quarto-search-button" aria-label="Search" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">2025 Coriell Bioinformatics Internship</a> 
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bioinformatics Notes</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-best-practices.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">‘Best’ practices for data projects</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./linux.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Command line basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./r-basics.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">R programming basics</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./data-viz.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Effective data visualizations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./presenting-science.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Scientific presentations</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./summarized-experiments.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">Practical introduction to SummarizedExperiments</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-expression-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Differential Expression Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./differential-methylation-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Differential Methylation Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./atacseq-analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">ATAC-seq Analysis</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./references.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">References</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#what-are-summarizedexperiments" id="toc-what-are-summarizedexperiments" class="nav-link active" data-scroll-target="#what-are-summarizedexperiments">What are <code>SummarizedExperiments</code></a>
  <ul class="collapse">
  <li><a href="#constructing-a-summarizedexperiment" id="toc-constructing-a-summarizedexperiment" class="nav-link" data-scroll-target="#constructing-a-summarizedexperiment">Constructing a <code>SummarizedExperiment</code></a></li>
  </ul></li>
  <li><a href="#accessing-parts-of-the-summarizedexperiment-object" id="toc-accessing-parts-of-the-summarizedexperiment-object" class="nav-link" data-scroll-target="#accessing-parts-of-the-summarizedexperiment-object">Accessing parts of the <code>SummarizedExperiment</code> object</a>
  <ul class="collapse">
  <li><a href="#getting-the-count-matrix" id="toc-getting-the-count-matrix" class="nav-link" data-scroll-target="#getting-the-count-matrix">Getting the count matrix</a></li>
  <li><a href="#getting-the-column-metadata" id="toc-getting-the-column-metadata" class="nav-link" data-scroll-target="#getting-the-column-metadata">Getting the column metadata</a></li>
  <li><a href="#getting-the-rowranges" id="toc-getting-the-rowranges" class="nav-link" data-scroll-target="#getting-the-rowranges">Getting the rowRanges</a></li>
  <li><a href="#getting-the-rowdata" id="toc-getting-the-rowdata" class="nav-link" data-scroll-target="#getting-the-rowdata">Getting the rowData</a></li>
  </ul></li>
  <li><a href="#modifying-a-summarizedexperiment" id="toc-modifying-a-summarizedexperiment" class="nav-link" data-scroll-target="#modifying-a-summarizedexperiment">Modifying a <code>SummarizedExperiment</code></a>
  <ul class="collapse">
  <li><a href="#adding-assays" id="toc-adding-assays" class="nav-link" data-scroll-target="#adding-assays">Adding assays</a></li>
  <li><a href="#adding-metadata" id="toc-adding-metadata" class="nav-link" data-scroll-target="#adding-metadata">Adding metadata</a></li>
  <li><a href="#adding-rowdata" id="toc-adding-rowdata" class="nav-link" data-scroll-target="#adding-rowdata">Adding rowData</a></li>
  </ul></li>
  <li><a href="#subsetting-summarizedexperiment-objects" id="toc-subsetting-summarizedexperiment-objects" class="nav-link" data-scroll-target="#subsetting-summarizedexperiment-objects">Subsetting <code>SummarizedExperiment</code> objects</a>
  <ul class="collapse">
  <li><a href="#subsetting-based-on-sample-metadata" id="toc-subsetting-based-on-sample-metadata" class="nav-link" data-scroll-target="#subsetting-based-on-sample-metadata">Subsetting based on sample metadata</a></li>
  <li><a href="#subsetting-based-on-rows" id="toc-subsetting-based-on-rows" class="nav-link" data-scroll-target="#subsetting-based-on-rows">Subsetting based on rows</a></li>
  <li><a href="#subsetting-based-on-rowranges" id="toc-subsetting-based-on-rowranges" class="nav-link" data-scroll-target="#subsetting-based-on-rowranges">Subsetting based on rowRanges</a></li>
  <li><a href="#combining-subsetting-operations" id="toc-combining-subsetting-operations" class="nav-link" data-scroll-target="#combining-subsetting-operations">Combining subsetting operations</a></li>
  </ul></li>
  <li><a href="#saving-a-summarizedexperiment" id="toc-saving-a-summarizedexperiment" class="nav-link" data-scroll-target="#saving-a-summarizedexperiment">Saving a <code>SummarizedExperiment</code></a></li>
  <li><a href="#summarizedexperiments-in-the-real-world" id="toc-summarizedexperiments-in-the-real-world" class="nav-link" data-scroll-target="#summarizedexperiments-in-the-real-world"><code>SummarizedExperiments</code> in the real world</a></li>
  <li><a href="#closing-thoughts" id="toc-closing-thoughts" class="nav-link" data-scroll-target="#closing-thoughts">Closing thoughts</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Practical introduction to SummarizedExperiments</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="what-are-summarizedexperiments" class="level2">
<h2 class="anchored" data-anchor-id="what-are-summarizedexperiments">What are <code>SummarizedExperiments</code></h2>
<p><code>SummarizedExperiments</code> are R objects meant for organizing and manipulating rectangular matrices that are typically produced by arrays or high-throughput sequencing. If you are doing any kind of analysis that requires associating feature-level data (RNA-seq gene counts, methylation array loci, ATAC-seq regions, etc.) with the genomic coordinates of those features and the sample-level metadata with which those features were measured, then you should be using a <code>SummarizedExperiment</code> to organize, manipulate, and store your results.</p>
<p>Please take a moment to read through the first 2 sections (at least) of the <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">SummarizedExperiment vignette</a> in order to familiarize yourself with what <code>SummarizedExperiments</code> are and their structure. I will demonstrate <em>how</em> you can use <code>SummarizedExperiments</code> below.</p>
<p>From the <code>SummarizedExperiment</code> <a href="https://bioconductor.org/packages/release/bioc/vignettes/SummarizedExperiment/inst/doc/SummarizedExperiment.html">vignette</a>:</p>
<p><img src="images/summarized-experiments.png" title="Overview of SummarizedExperiment structure" class="img-fluid"></p>
<p>The <code>SummarizedExperiment</code> object coordinates four main parts:</p>
<ul>
<li><code>assay()</code>, <code>assays()</code>: A matrix-like or list of matrix-like objects of identical dimension
<ul>
<li>matrix-like: implements <code>dim()</code>, <code>dimnames()</code>, and 2-dimensional <code>[</code>, <code>[&lt;-</code> methods.</li>
<li>rows: genes, genomic coordinates, etc.</li>
<li>columns: samples, cells, etc.</li>
</ul></li>
<li><code>colData()</code>: Annotations on each column, as a DataFrame.
<ul>
<li>E.g., description of each sample</li>
</ul></li>
<li><code>rowData()</code> and / or <code>rowRanges()</code>: Annotations on each row.
<ul>
<li><code>rowRanges()</code>: coordinates of gene / exons in transcripts / etc.</li>
<li><code>rowData()</code>: P-values and log-fold change of each gene after differential expression analysis.</li>
</ul></li>
<li><code>metadata()</code>: List of unstructured metadata describing the overall content of the object.</li>
</ul>
<p>In order to better understand how they work, let’s construct a <code>SummarizedExperiment</code> from scratch.</p>
<section id="constructing-a-summarizedexperiment" class="level3">
<h3 class="anchored" data-anchor-id="constructing-a-summarizedexperiment">Constructing a <code>SummarizedExperiment</code></h3>
<p>Hopefully you’ll already be working with data that is in a <code>SummarizedExperiment</code> or some other class that derives from one. But just in case you don’t have data structured as a <code>SummarizedExperiment</code> it’s useful and instructive to understand how to create one from scratch.</p>
<p>To be most useful, a <code>SummarizedExperiment</code> should have at least:</p>
<ul>
<li>A matrix of data with features in rows and samples in columns</li>
<li>A metadata data.frame with samples as rownames and columns describing their properties</li>
</ul>
<p>Another really useful object to add to <code>SummarizedExperiments</code> is a GRanges object describing the genomic locations of each feature in the matrix. Adding this to the <code>SummarizedExperiment</code> creates what is called a <code>RangedSummarizedExperiment</code> that acts just like a regular <code>SummarizedExperiment</code> with some extra features.</p>
<p>To construct our basic <code>SummarizedExperiment</code>:</p>
<ul>
<li>We’ll create a ‘counts’ matrix with gene IDs as rows and Samples in columns</li>
<li>We’ll add some metadata describing the Samples</li>
<li>We’ll add on <code>GRanges()</code> describing the genomic location of the genes</li>
</ul>
<section id="construct-the-counts-matrix" class="level4">
<h4 class="anchored" data-anchor-id="construct-the-counts-matrix">Construct the counts matrix</h4>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">library</span>(SummarizedExperiment))</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">matrix</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">data =</span> <span class="fu">rnbinom</span>(<span class="at">n =</span> <span class="dv">200</span> <span class="sc">*</span> <span class="dv">6</span>, <span class="at">mu =</span> <span class="dv">100</span>, <span class="at">size =</span> <span class="dv">1</span> <span class="sc">/</span> <span class="fl">0.5</span>),</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">nrow =</span> <span class="dv">200</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">dimnames =</span> <span class="fu">list</span>(<span class="fu">paste0</span>(<span class="st">"Gene"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>), <span class="fu">paste0</span>(<span class="st">"Sample"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>))</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a>counts[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2 Sample3 Sample4 Sample5
Gene1      43     195     205      55      73
Gene2      37     102      23     130      36
Gene3      73      60      33      46      34
Gene4     119     123     346      26     149
Gene5     121      38     230     132     317</code></pre>
</div>
</div>
</section>
<section id="construct-the-sample-metadata" class="level4">
<h4 class="anchored" data-anchor-id="construct-the-sample-metadata">Construct the sample metadata</h4>
<p>It is important that the sample metadata be either a <code>data.frame</code> or a <code>DataFrame</code> object because <code>SummarizedExperiment</code> requires the <code>colData()</code> to have rownames that match the colnames of the count matrix.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>coldata <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">SampleName =</span> <span class="fu">colnames</span>(counts),</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Treatment =</span> <span class="fu">gl</span>(<span class="dv">2</span>, <span class="dv">3</span>, <span class="at">labels =</span> <span class="fu">c</span>(<span class="st">"Control"</span>, <span class="st">"Treatment"</span>)),</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">Age =</span> <span class="fu">sample.int</span>(<span class="dv">100</span>, <span class="dv">6</span>),</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">row.names =</span> <span class="fu">colnames</span>(counts)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>coldata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        SampleName Treatment Age
Sample1    Sample1   Control  32
Sample2    Sample2   Control  96
Sample3    Sample3   Control  41
Sample4    Sample4 Treatment  48
Sample5    Sample5 Treatment  52
Sample6    Sample6 Treatment   2</code></pre>
</div>
</div>
<p>Notice that all of the rownames of the metadata are in the same order as the colnames of the counts matrix. <strong>This is necessary</strong>.</p>
</section>
<section id="construct-gene-range-annotations" class="level4">
<h4 class="anchored" data-anchor-id="construct-gene-range-annotations">Construct gene range annotations</h4>
<p>You will usually have gene annotations or <code>GRanges</code> objects loaded from a GTF file or you may even create <code>GRanges</code> yourself by specifying the chromosome, start, end, and strand, information manually.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>rowranges <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"chr1"</span>, <span class="st">"chr2"</span>), <span class="fu">c</span>(<span class="dv">50</span>, <span class="dv">150</span>)),</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">IRanges</span>(<span class="fu">floor</span>(<span class="fu">runif</span>(<span class="dv">200</span>, <span class="fl">1e5</span>, <span class="fl">1e6</span>)), <span class="at">width =</span> <span class="dv">100</span>),</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">strand =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"+"</span>, <span class="st">"-"</span>), <span class="dv">200</span>, <span class="cn">TRUE</span>),</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">feature_id =</span> <span class="fu">sprintf</span>(<span class="st">"ID%03d"</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">200</span>),</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">gene_type =</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">"protein_coding"</span>, <span class="st">"lncRNA"</span>, <span class="st">"repeat_element"</span>), <span class="dv">200</span>, <span class="at">replace =</span> <span class="cn">TRUE</span>)</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(rowranges) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(counts)</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Take a peek at what this looks like</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>rowranges</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 2 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt;
    Gene1     chr1 257333-257432      - |       ID001 protein_coding
    Gene2     chr1 543028-543127      + |       ID002 protein_coding
    Gene3     chr1 463891-463990      + |       ID003 protein_coding
    Gene4     chr1 679135-679234      + |       ID004         lncRNA
    Gene5     chr1 829784-829883      - |       ID005 protein_coding
      ...      ...           ...    ... .         ...            ...
  Gene196     chr2 723767-723866      - |       ID196         lncRNA
  Gene197     chr2 673554-673653      + |       ID197 repeat_element
  Gene198     chr2 593478-593577      - |       ID198 repeat_element
  Gene199     chr2 543982-544081      + |       ID199         lncRNA
  Gene200     chr2 114505-114604      - |       ID200 protein_coding
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="construct-the-summarizedexperiment-object" class="level4">
<h4 class="anchored" data-anchor-id="construct-the-summarizedexperiment-object">Construct the <code>SummarizedExperiment</code> object</h4>
<p>With these pieces of information we’re ready to create a <code>SummarizedExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>se <span class="ot">&lt;-</span> <span class="fu">SummarizedExperiment</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">assays =</span> <span class="fu">list</span>(<span class="at">counts =</span> counts),</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">colData =</span> coldata,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">rowRanges =</span> rowranges</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Printing the object gives a summary of what's inside</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>se</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 6 
metadata(0):
assays(1): counts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(2): feature_id gene_type
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(3): SampleName Treatment Age</code></pre>
</div>
</div>
</section>
</section>
</section>
<section id="accessing-parts-of-the-summarizedexperiment-object" class="level2">
<h2 class="anchored" data-anchor-id="accessing-parts-of-the-summarizedexperiment-object">Accessing parts of the <code>SummarizedExperiment</code> object</h2>
<p>Every part of the <code>SummarizedExperiment</code> object can be extracted with its accessor function. To extract a particular assay you can use the <code>assay()</code> function. To extract the column metadata you can use the <code>colData()</code> function. To extract the GRanges for the rows of the matrix you can use the <code>rowRanges()</code> function. The <code>rowData()</code> function also allows you to access row-level annotation information from data added to the <code>rowData</code> slot or by the <code>mcols()</code> of the <code>rowRanges</code>. This will be made more clear below.</p>
<section id="getting-the-count-matrix" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-count-matrix">Getting the count matrix</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"counts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
Gene1      43     195     205      55      73     105
Gene2      37     102      23     130      36      91
Gene3      73      60      33      46      34     105
Gene4     119     123     346      26     149      52
Gene5     121      38     230     132     317      63
Gene6      14      30      73      45     121      86</code></pre>
</div>
</div>
<p>To see what assays are available you can use the <code>assays()</code> function</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 1
names(1): counts</code></pre>
</div>
</div>
</section>
<section id="getting-the-column-metadata" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-column-metadata">Getting the column metadata</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 3 columns
         SampleName Treatment       Age
        &lt;character&gt;  &lt;factor&gt; &lt;integer&gt;
Sample1     Sample1 Control          32
Sample2     Sample2 Control          96
Sample3     Sample3 Control          41
Sample4     Sample4 Treatment        48
Sample5     Sample5 Treatment        52
Sample6     Sample6 Treatment         2</code></pre>
</div>
</div>
</section>
<section id="getting-the-rowranges" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-rowranges">Getting the rowRanges</h3>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 200 ranges and 2 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt;
    Gene1     chr1 257333-257432      - |       ID001 protein_coding
    Gene2     chr1 543028-543127      + |       ID002 protein_coding
    Gene3     chr1 463891-463990      + |       ID003 protein_coding
    Gene4     chr1 679135-679234      + |       ID004         lncRNA
    Gene5     chr1 829784-829883      - |       ID005 protein_coding
      ...      ...           ...    ... .         ...            ...
  Gene196     chr2 723767-723866      - |       ID196         lncRNA
  Gene197     chr2 673554-673653      + |       ID197 repeat_element
  Gene198     chr2 593478-593577      - |       ID198 repeat_element
  Gene199     chr2 543982-544081      + |       ID199         lncRNA
  Gene200     chr2 114505-114604      - |       ID200 protein_coding
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
</section>
<section id="getting-the-rowdata" class="level3">
<h3 class="anchored" data-anchor-id="getting-the-rowdata">Getting the rowData</h3>
<p>Note that rowData in this case is the same as <code>mcols()</code> of the <code>rowRanges</code></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 200 rows and 2 columns
         feature_id      gene_type
        &lt;character&gt;    &lt;character&gt;
Gene1         ID001 protein_coding
Gene2         ID002 protein_coding
Gene3         ID003 protein_coding
Gene4         ID004         lncRNA
Gene5         ID005 protein_coding
...             ...            ...
Gene196       ID196         lncRNA
Gene197       ID197 repeat_element
Gene198       ID198 repeat_element
Gene199       ID199         lncRNA
Gene200       ID200 protein_coding</code></pre>
</div>
</div>
</section>
</section>
<section id="modifying-a-summarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="modifying-a-summarizedexperiment">Modifying a <code>SummarizedExperiment</code></h2>
<p>Once you create a <code>SummarizedExperiment</code> you are not stuck with the information in that object. <code>SummarizedExperiments</code> allow you to add and modify the data within the object.</p>
<section id="adding-assays" class="level3">
<h3 class="anchored" data-anchor-id="adding-assays">Adding assays</h3>
<p>For example, we may wish to calculate counts per million values for our counts matrix and add a new assay back into our <code>SummarizedExperiment</code> object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate counts per million</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>counts <span class="ot">&lt;-</span> <span class="fu">assay</span>(se, <span class="st">"counts"</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a>cpm <span class="ot">&lt;-</span> counts <span class="sc">/</span> <span class="fu">colSums</span>(counts) <span class="sc">*</span> <span class="fl">1e6</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Add the CPM data as a new assay to our existing se object</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> cpm</span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a><span class="co"># And if we wish to log-scale these values</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"logcounts"</span>) <span class="ot">&lt;-</span> <span class="fu">log2</span>(cpm)</span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Now there are three assays available</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 3
names(3): counts cpm logcounts</code></pre>
</div>
</div>
<p>Note: Instead of creating intermediate variables we could also directly use the assays like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> <span class="fu">assay</span>(se, <span class="st">"counts"</span>) <span class="sc">/</span> <span class="fu">colSums</span>(<span class="fu">assay</span>(se, <span class="st">"counts"</span>)) <span class="sc">*</span> <span class="fl">1e6</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="adding-metadata" class="level3">
<h3 class="anchored" data-anchor-id="adding-metadata">Adding metadata</h3>
<p><code>SummarizedExperiment</code> objects use the <code>$</code> to get and set columns of the metadata contained in the <code>colData</code> slot. For example, to get all of the Ages we can use:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>se<span class="sc">$</span>Age</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 32 96 41 48 52  2</code></pre>
</div>
</div>
<p>If we want to add a new column we simply create the new column in the same way</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>se<span class="sc">$</span>Batch <span class="ot">&lt;-</span> <span class="fu">factor</span>(<span class="fu">rep</span>(<span class="fu">c</span>(<span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>), <span class="dv">2</span>))</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Now you can se that a new 'Batch` column has been added to the colData</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 6 rows and 4 columns
         SampleName Treatment       Age    Batch
        &lt;character&gt;  &lt;factor&gt; &lt;integer&gt; &lt;factor&gt;
Sample1     Sample1 Control          32        A
Sample2     Sample2 Control          96        B
Sample3     Sample3 Control          41        C
Sample4     Sample4 Treatment        48        A
Sample5     Sample5 Treatment        52        B
Sample6     Sample6 Treatment         2        C</code></pre>
</div>
</div>
</section>
<section id="adding-rowdata" class="level3">
<h3 class="anchored" data-anchor-id="adding-rowdata">Adding rowData</h3>
<p>We can also modify the data which describes each feature in the matrix by adding columns to the rowData. For example, let’s create a new column called <code>Keep</code> if the gene is a protein_coding gene.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)<span class="sc">$</span>Keep <span class="ot">&lt;-</span> <span class="fu">rowData</span>(se)<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"protein_coding"</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 200 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene1         ID001 protein_coding      TRUE
Gene2         ID002 protein_coding      TRUE
Gene3         ID003 protein_coding      TRUE
Gene4         ID004         lncRNA     FALSE
Gene5         ID005 protein_coding      TRUE
...             ...            ...       ...
Gene196       ID196         lncRNA     FALSE
Gene197       ID197 repeat_element     FALSE
Gene198       ID198 repeat_element     FALSE
Gene199       ID199         lncRNA     FALSE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
</section>
</section>
<section id="subsetting-summarizedexperiment-objects" class="level2">
<h2 class="anchored" data-anchor-id="subsetting-summarizedexperiment-objects">Subsetting <code>SummarizedExperiment</code> objects</h2>
<p><code>SummarizedExperiments</code> follow the basic idea of</p>
<pre><code>se[the rows you want, the columns you want]</code></pre>
<p>With a <code>SummarizedExperiment</code> “the rows you want” corresponds to the features in the rows of the matrix/rowData and “the columns you want” corresponds to the metadata in <code>colData</code></p>
<section id="subsetting-based-on-sample-metadata" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-based-on-sample-metadata">Subsetting based on sample metadata</h3>
<p>For example, if we want to select all of the data belonging only to samples in the Treatment group we can use the following:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>(trt <span class="ot">&lt;-</span> se[, se<span class="sc">$</span>Treatment <span class="sc">==</span> <span class="st">"Treatment"</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 3 
metadata(0):
assays(3): counts cpm logcounts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(3): Sample4 Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>Notice how the <code>dim</code> of the object changed from 6 to 3. This is because we have selected only the Samples from the original <code>SummarizedExperiment</code> object from the treatment group. The cool thing about <code>SummarizedExperiments</code> is that all of the assays have also been subsetted to reflect this selection!</p>
<p>Take a look at the “logcounts” assay. It only contains Samples 4, 5, and 6.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(trt, <span class="st">"logcounts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Sample4  Sample5  Sample6
Gene1 11.44995 11.80667 12.28449
Gene2 12.58495 10.90951 12.07571
Gene3 11.14041 10.65770 12.38284
Gene4 10.44002 12.78708 11.26302
Gene5 12.61464 13.97693 11.59412
Gene6 11.05977 12.48145 12.16585</code></pre>
</div>
</div>
<p>Of course you can combine multiple conditions as well</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>se[, se<span class="sc">$</span>Batch <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"B"</span>, <span class="st">"C"</span>) <span class="sc">&amp;</span> se<span class="sc">$</span>Age <span class="sc">&gt;</span> <span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 200 3 
metadata(0):
assays(3): counts cpm logcounts
rownames(200): Gene1 Gene2 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(3): Sample2 Sample3 Sample5
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
</section>
<section id="subsetting-based-on-rows" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-based-on-rows">Subsetting based on rows</h3>
<p>We can also select certain features that we want to keep using row subsetting. For example to select only the first 50 rows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>se[<span class="dv">1</span><span class="sc">:</span><span class="dv">50</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 50 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(50): Gene1 Gene2 ... Gene49 Gene50
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>Notice how the <code>dim</code> changed from 200 to 50 reflecting the fact that we have only selected the first 50 rows.</p>
<p>This subsetting is very useful when combined with logical operators. Above we created a vector in <code>rowData</code> called “Keep” that contained TRUE if the corresponding row of the se object was a coding gene and FALSE otherwise. Let’s use this vector to subset our se object.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>(coding <span class="ot">&lt;-</span> se[<span class="fu">rowData</span>(se)<span class="sc">$</span>Keep, ])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 61 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(61): Gene1 Gene2 ... Gene194 Gene200
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>And if we look at the resulting <code>rowData</code> we can see that it only contains the protein_coding features</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(coding)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 61 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene1         ID001 protein_coding      TRUE
Gene2         ID002 protein_coding      TRUE
Gene3         ID003 protein_coding      TRUE
Gene5         ID005 protein_coding      TRUE
Gene13        ID013 protein_coding      TRUE
...             ...            ...       ...
Gene188       ID188 protein_coding      TRUE
Gene191       ID191 protein_coding      TRUE
Gene193       ID193 protein_coding      TRUE
Gene194       ID194 protein_coding      TRUE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
<p>Each assay also reflects this operation</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(coding, <span class="st">"cpm"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>        Sample1  Sample2   Sample3  Sample4   Sample5  Sample6
Gene1  2187.182 9569.143  9740.105 2797.558  3582.295 4988.835
Gene2  1748.665 5449.882  1091.030 6143.958  1923.488 4316.683
Gene3  3582.295 2850.763  1678.535 2257.336  1615.432 5340.793
Gene5  5749.038 1932.859 11286.682 6271.678 16124.110 3091.569
Gene13 1678.535 5152.616  1140.305 5645.982  4858.180 2043.047
Gene15 2748.062 7459.495  4323.499 3484.150  3658.479 3102.747</code></pre>
</div>
</div>
</section>
<section id="subsetting-based-on-rowranges" class="level3">
<h3 class="anchored" data-anchor-id="subsetting-based-on-rowranges">Subsetting based on rowRanges</h3>
<p>A closely related row-wise subsetting operation can be used if you have a <code>RangedSummarizedExperiment</code> (a <code>SummarizedExperiment</code> with a rowRanges slot) that allows you to perform operations on a <code>SummarizedExperiment</code> object like you would a GRanges object.</p>
<p>For example, let’s say we only wanted to extract the features on Chromosome 2 only. Then we can use the <code>GenomicRanges</code> function <code>subsetByOverlaps</code> directly on our <code>SummarizedExperiment</code> object like so:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Region of interest</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>roi <span class="ot">&lt;-</span> <span class="fu">GRanges</span>(<span class="at">seqnames =</span> <span class="st">"chr2"</span>, <span class="at">ranges =</span> <span class="dv">1</span><span class="sc">:</span><span class="fl">1e7</span>)</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the SE object for only features on chr2</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>(chr2 <span class="ot">&lt;-</span> <span class="fu">subsetByOverlaps</span>(se, roi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 150 6 
metadata(0):
assays(3): counts cpm logcounts
rownames(150): Gene51 Gene52 ... Gene199 Gene200
rowData names(3): feature_id gene_type Keep
colnames(6): Sample1 Sample2 ... Sample5 Sample6
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
<p>You can see again that the <code>dim</code> changed reflecting our selection. Again, all of the associated assays and rowData have also been subsetted reflecting this change as well.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(chr2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 150 rows and 3 columns
         feature_id      gene_type      Keep
        &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
Gene51        ID051 protein_coding      TRUE
Gene52        ID052 protein_coding      TRUE
Gene53        ID053 repeat_element     FALSE
Gene54        ID054 repeat_element     FALSE
Gene55        ID055 protein_coding      TRUE
...             ...            ...       ...
Gene196       ID196         lncRNA     FALSE
Gene197       ID197 repeat_element     FALSE
Gene198       ID198 repeat_element     FALSE
Gene199       ID199         lncRNA     FALSE
Gene200       ID200 protein_coding      TRUE</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(chr2, <span class="st">"counts"</span>) <span class="sc">|&gt;</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       Sample1 Sample2 Sample3 Sample4 Sample5 Sample6
Gene51     231     173      57      66     138      48
Gene52      40     120      60     126     223      80
Gene53      47     155     109      54     138      48
Gene54     107      95      93     120     134     198
Gene55      89      33      87      67     171      85
Gene56      35      65      96      61     160      95</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowRanges</span>(chr2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>GRanges object with 150 ranges and 3 metadata columns:
          seqnames        ranges strand |  feature_id      gene_type      Keep
             &lt;Rle&gt;     &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;    &lt;character&gt; &lt;logical&gt;
   Gene51     chr2 199333-199432      + |       ID051 protein_coding      TRUE
   Gene52     chr2 405579-405678      + |       ID052 protein_coding      TRUE
   Gene53     chr2 753001-753100      - |       ID053 repeat_element     FALSE
   Gene54     chr2 644644-644743      + |       ID054 repeat_element     FALSE
   Gene55     chr2 115245-115344      - |       ID055 protein_coding      TRUE
      ...      ...           ...    ... .         ...            ...       ...
  Gene196     chr2 723767-723866      - |       ID196         lncRNA     FALSE
  Gene197     chr2 673554-673653      + |       ID197 repeat_element     FALSE
  Gene198     chr2 593478-593577      - |       ID198 repeat_element     FALSE
  Gene199     chr2 543982-544081      + |       ID199         lncRNA     FALSE
  Gene200     chr2 114505-114604      - |       ID200 protein_coding      TRUE
  -------
  seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
</div>
<p>There’s also a few shortcuts on range operations using GRanges/SummarizedExperiments. See the help pages for <code>%over</code>, <code>%within%</code>, and <code>%outside%</code>. For example:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all.equal</span>(se[se <span class="sc">%over%</span> roi, ], <span class="fu">subsetByOverlaps</span>(se, roi))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
</section>
<section id="combining-subsetting-operations" class="level3">
<h3 class="anchored" data-anchor-id="combining-subsetting-operations">Combining subsetting operations</h3>
<p>Of course you don’t have to perform one subsetting operation at a time. Like base R you can combine multiple expressions to subset a <code>SummarizedExperiment</code> object.</p>
<p>For example, to select only features labeled as repeat_elements and the Sample from ‘Batch’ A in the ‘Control’ group</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>(selected <span class="ot">&lt;-</span> se[</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rowData</span>(se)<span class="sc">$</span>gene_type <span class="sc">==</span> <span class="st">"repeat_element"</span>,</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  se<span class="sc">$</span>Treatment <span class="sc">==</span> <span class="st">"Control"</span> <span class="sc">&amp;</span></span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>    se<span class="sc">$</span>Batch <span class="sc">==</span> <span class="st">"A"</span></span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: RangedSummarizedExperiment 
dim: 62 1 
metadata(0):
assays(3): counts cpm logcounts
rownames(62): Gene8 Gene9 ... Gene197 Gene198
rowData names(3): feature_id gene_type Keep
colnames(1): Sample1
colData names(4): SampleName Treatment Age Batch</code></pre>
</div>
</div>
</section>
</section>
<section id="saving-a-summarizedexperiment" class="level2">
<h2 class="anchored" data-anchor-id="saving-a-summarizedexperiment">Saving a <code>SummarizedExperiment</code></h2>
<p>Since <code>SummarizedExperiments</code> keep basically all information about an experiment in one place, it is convenient to save the entire <code>SummarizedExperiment</code> object so that you can pick up an analysis where you left off or even to facilitate better sharing of data between collaborators.</p>
<p>You can save the entire <code>SummarizedExperiment</code> object with:</p>
<p><code>saveRDS(se, "/path/to/se.rds")</code></p>
<p>And when you want to read the same object back into R for your next analysis you can do so with:</p>
<p><code>se &lt;- readRDS("/path/to/se.rds")</code></p>
</section>
<section id="summarizedexperiments-in-the-real-world" class="level2">
<h2 class="anchored" data-anchor-id="summarizedexperiments-in-the-real-world"><code>SummarizedExperiments</code> in the real world</h2>
<p>If you’re working with any <a href="https://bioconductor.org/">Bioconductor</a> packages it’s likely that the object you’re working with either is a <code>SummarizedExperiment</code> or is inherited from one. For example, the <code>DESeqDataSet</code> from the <a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2 package</a> and <code>BSseq</code> objects from the <a href="https://bioconductor.org/packages/release/bioc/html/bsseq.html">bsseq package</a> both inherit from a <code>SummarizedExperiment</code> and thus retain all of the same functionality as above. If you go to the <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> landing page and click “See More” under details you can see all of the packages that depend on <code>SummarizedExperiment</code>.</p>
<p>Also, many common methods are also implemented for <code>SummarizedExperiment</code> objects. For example, to simplify calculating counts-per-million above I could have simply used the <code>edgeR::cpm()</code> directly on the <code>SummarizedExperiment</code> object. Many functions in bioconductor packages know how to deal directly with <code>SummarizedExperiments</code> so you don’t ever have to take the trouble extracting components and performing tedious calculations yourself.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(se, <span class="st">"cpm"</span>) <span class="ot">&lt;-</span> edgeR<span class="sc">::</span><span class="fu">cpm</span>(se)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>I also left out any discussion of the <code>metadata()</code> slot of the <code>SummarizedExperiment</code>. The metadata slot is simply a list of any R object that contains information about the experiment. The <code>metadata</code> in the metadata slots are not subjected to the same subsetting rules as the other slots. In practice this assay contains additional information about the experiment as a whole. For example, I typically store bootstrap alignments for each sample here.</p>
<p>To add something to the <code>SummarizedExperiment</code> metadata slot you can do:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(se)<span class="sc">$</span>additional_info <span class="ot">&lt;-</span> <span class="st">"Experiment performed on 6 samples with three replicates each"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>And to retrieve this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="fu">metadata</span>(se)<span class="sc">$</span>additional_info</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Experiment performed on 6 samples with three replicates each"</code></pre>
</div>
</div>
</section>
<section id="closing-thoughts" class="level2">
<h2 class="anchored" data-anchor-id="closing-thoughts">Closing thoughts</h2>
<p>Hopefully this was enough information to get you started using <code>SummarizedExperiments</code>. There’s many things I left out such as different <a href="https://bioconductor.org/packages/release/bioc/html/HDF5Array.html">backings</a> for storing out of memory data, a tidyverse <a href="https://stemangiola.github.io/tidySummarizedExperiment/">interface</a> to <code>SummarizedExperiment</code> objects, <a href="https://bioconductor.org/packages/release/bioc/vignettes/mia/inst/doc/mia.html">TreeSummarizedExperiments</a> for microbiome data, <a href="https://bioconductor.org/packages/release/bioc/vignettes/MultiAssayExperiment/inst/doc/MultiAssayExperiment.html">MultiAssayExperiments</a> for dealing with experiments containing multiomics data, and much more.</p>
<p>Please let me know your thoughts and if anything needs more clarification.</p>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./presenting-science.html" class="pagination-link" aria-label="Scientific presentations">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text">Scientific presentations</span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./differential-expression-analysis.html" class="pagination-link" aria-label="Differential Expression Analysis">
        <span class="nav-page-text">Differential Expression Analysis</span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->




</body></html>